{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "40f86f79",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘Rsubread’ was built under R version 4.2.1”\n",
      "Warning message:\n",
      "“package ‘qqman’ was built under R version 4.2.1”\n",
      "\n",
      "\n",
      "For example usage please run: vignette('qqman')\n",
      "\n",
      "\n",
      "\n",
      "Citation appreciated but not required:\n",
      "\n",
      "Turner, (2018). qqman: an R package for visualizing GWAS results using Q-Q and manhattan plots. Journal of Open Source Software, 3(25), 731, https://doi.org/10.21105/joss.00731.\n",
      "\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘DESeq2’ was built under R version 4.2.2”\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Warning message:\n",
      "“package ‘BiocGenerics’ was built under R version 4.2.1”\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘IRanges’ was built under R version 4.2.1”\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomicRanges’ was built under R version 4.2.2”\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomeInfoDb’ was built under R version 4.2.2”\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Warning message:\n",
      "“package ‘SummarizedExperiment’ was built under R version 4.2.2”\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Warning message:\n",
      "“package ‘MatrixGenerics’ was built under R version 4.2.1”\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Warning message:\n",
      "“package ‘matrixStats’ was built under R version 4.2.2”\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Warning message:\n",
      "“package ‘Biobase’ was built under R version 4.2.1”\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘gridExtra’ was built under R version 4.2.1”\n",
      "\n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘circlize’ was built under R version 4.2.1”\n",
      "========================================\n",
      "circlize version 0.4.15\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# loading libraries:\n",
    "library(\"ggplot2\")\n",
    "library(\"Rsubread\") # featureCount\n",
    "library(\"stats\") # for PCA\n",
    "library(\"qqman\") # qq-plot, manhattan plot\n",
    "library(\"DESeq2\") # for rlog data\n",
    "library(\"gridExtra\") # several ggplots in one pdf\n",
    "library(\"circlize\") # colorRamp2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "99956e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/data/bcu_projects/MelBrainSys_PostdocProject_Gruetzmann/github/RNA-Seq-analysis-example/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d13ad0c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm(list=ls())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7198cf97",
   "metadata": {},
   "outputs": [],
   "source": [
    "imageMagickConvert = \"/home/gruetzmko/miniconda3/envs/imagemagick/bin/convert\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "af75b50e",
   "metadata": {},
   "outputs": [],
   "source": [
    "padjCutoff <- 0.05\n",
    "pCutoff <- 0.05\n",
    "RNAScriptPath = \n",
    " \"sftp://gruetzmko@sv69-comp1.ukd69.med.tu-dresden.de/data/bcu_projects/MelBrainSys_PostdocProject_Gruetzmann/github/RNA-Seq-analysis-example/scripts/\"\n",
    "\n",
    "args = c(\"diff-gene-exp\",\n",
    "            \"annotation/GRCh37.p23-EnsgID-GeneSymbol.csv.gz\",\n",
    "           \"$RNAScriptPath/deseq-results-template.xlsx\",\n",
    "           \"annotation/annotation.csv\", \"featCount-noMultiMap.RData\", \n",
    "            \"genetypes-remove.txt\",\n",
    "           \"diff-gene-exp/deseqComparisons.txt\",\n",
    "           \"diff-gene-exp/DGE-statistics.csv\",\n",
    "           \"n\",\"y\",\"10\",\"0\",\"0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "449872d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# args <- commandArgs(trailingOnly = T)\n",
    "if(length(args) !=13) {\n",
    "  cat(\"ERROR: 13 parameters needed: outDir, ensembl gene descr, deseq result table template (template.xlsx), sample annot,\n",
    "      featureCountFile, file with genetypes to include/exclude, txt file with deseq comparisons to make, \n",
    "      overviewStatFile to write, file with EnsgIDs to keep extra (or n),\n",
    "      save all results (y/n), min reads per gene in sum, min reads per gene and sample,\n",
    "      min read counts for min 2 samples\\nyou had\",length(args),\" (\",paste(args,collapse=\", \"),\")\\n\")\n",
    "  # q(\"no\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0faa7e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "outDir <- paste(args[1],\"/\",sep=\"\")\n",
    "geneDescrFile <- args[2]\n",
    "deseqTableTemplate <- args[3]\n",
    "sampleAnnotFile <- args[4]\n",
    "featureCountFile <- args[5]\n",
    "genetypeRemoveFile <- args[6]\n",
    "deseqCompareFile <- args[7]\n",
    "overviewStatFile <- args[8]\n",
    "extraEnsgIDFile <- args[9]\n",
    "saveAllRes <- args[10]\n",
    "minReadCountSum <- as.numeric(args[11])\n",
    "minReadCountPerSample <- as.numeric(args[12])\n",
    "minReadCountFor2Samples <- as.numeric(args[13])\n",
    "xlsLetters <- c(LETTERS,as.character(sapply(c(\"A\",\"B\",\"C\"),function(l)paste(l,LETTERS,sep=\"\"))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "47999b68",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (length(grep(pattern = \"y|n\",x = saveAllRes,ignore.case = T, perl =T))==0) {\n",
    "  cat(\"parameter 10 must be y or n (save all results)\\n\")\n",
    "  #q(\"no\")\n",
    "}\n",
    "saveAllRes <- ifelse(saveAllRes==\"y\",T,F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "eab21f00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Ensembl.Gene.ID</th><th scope=col>Associated.Gene.Name</th><th scope=col>Description</th><th scope=col>Gene.type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000271782</td><td>RP5-850O15.4</td><td>                                                                     </td><td>lincRNA  </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000232753</td><td>RP11-347K2.1</td><td>                                                                     </td><td>lincRNA  </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENSG00000225767</td><td>RP5-850O15.3</td><td>                                                                     </td><td>antisense</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENSG00000202140</td><td>Y_RNA       </td><td>Y RNA [Source:RFAM;Acc:RF00019]                                      </td><td>misc_RNA </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENSG00000207194</td><td>RNU6-1026P  </td><td>RNA, U6 small nuclear 1026, pseudogene [Source:HGNC Symbol;Acc:47989]</td><td>snRNA    </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENSG00000252825</td><td>RNU6-1253P  </td><td>RNA, U6 small nuclear 1253, pseudogene [Source:HGNC Symbol;Acc:48216]</td><td>snRNA    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & Ensembl.Gene.ID & Associated.Gene.Name & Description & Gene.type\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000271782 & RP5-850O15.4 &                                                                       & lincRNA  \\\\\n",
       "\t2 & ENSG00000232753 & RP11-347K2.1 &                                                                       & lincRNA  \\\\\n",
       "\t3 & ENSG00000225767 & RP5-850O15.3 &                                                                       & antisense\\\\\n",
       "\t4 & ENSG00000202140 & Y\\_RNA        & Y RNA {[}Source:RFAM;Acc:RF00019{]}                                       & misc\\_RNA \\\\\n",
       "\t5 & ENSG00000207194 & RNU6-1026P   & RNA, U6 small nuclear 1026, pseudogene {[}Source:HGNC Symbol;Acc:47989{]} & snRNA    \\\\\n",
       "\t6 & ENSG00000252825 & RNU6-1253P   & RNA, U6 small nuclear 1253, pseudogene {[}Source:HGNC Symbol;Acc:48216{]} & snRNA    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| <!--/--> | Ensembl.Gene.ID &lt;chr&gt; | Associated.Gene.Name &lt;chr&gt; | Description &lt;chr&gt; | Gene.type &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | ENSG00000271782 | RP5-850O15.4 | <!----> | lincRNA   |\n",
       "| 2 | ENSG00000232753 | RP11-347K2.1 | <!----> | lincRNA   |\n",
       "| 3 | ENSG00000225767 | RP5-850O15.3 | <!----> | antisense |\n",
       "| 4 | ENSG00000202140 | Y_RNA        | Y RNA [Source:RFAM;Acc:RF00019]                                       | misc_RNA  |\n",
       "| 5 | ENSG00000207194 | RNU6-1026P   | RNA, U6 small nuclear 1026, pseudogene [Source:HGNC Symbol;Acc:47989] | snRNA     |\n",
       "| 6 | ENSG00000252825 | RNU6-1253P   | RNA, U6 small nuclear 1253, pseudogene [Source:HGNC Symbol;Acc:48216] | snRNA     |\n",
       "\n"
      ],
      "text/plain": [
       "  Ensembl.Gene.ID Associated.Gene.Name\n",
       "1 ENSG00000271782 RP5-850O15.4        \n",
       "2 ENSG00000232753 RP11-347K2.1        \n",
       "3 ENSG00000225767 RP5-850O15.3        \n",
       "4 ENSG00000202140 Y_RNA               \n",
       "5 ENSG00000207194 RNU6-1026P          \n",
       "6 ENSG00000252825 RNU6-1253P          \n",
       "  Description                                                          \n",
       "1                                                                      \n",
       "2                                                                      \n",
       "3                                                                      \n",
       "4 Y RNA [Source:RFAM;Acc:RF00019]                                      \n",
       "5 RNA, U6 small nuclear 1026, pseudogene [Source:HGNC Symbol;Acc:47989]\n",
       "6 RNA, U6 small nuclear 1253, pseudogene [Source:HGNC Symbol;Acc:48216]\n",
       "  Gene.type\n",
       "1 lincRNA  \n",
       "2 lincRNA  \n",
       "3 antisense\n",
       "4 misc_RNA \n",
       "5 snRNA    \n",
       "6 snRNA    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "annot <- read.csv(file = gzfile(geneDescrFile),header = T,sep=\"\\t\",stringsAsFactors=F)\n",
    "head(annot)\n",
    "colnames(annot) <- c(\"ensgID\",\"symbol\",\"description\",\"geneType\")\n",
    "rownames(annot) <- annot$ensgID\n",
    "annot$description <- gsub(\"([^\\\\[]+).*\",\"\\\\1\",annot$description, perl=T)\n",
    "annot$description <- gsub(\"(.+) \",\"\\\\1\",annot$description,perl=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ecd1d3c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene type inclusion/exclusion :\n",
    "wishedGenetypes <- read.csv(genetypeRemoveFile,header=F,stringsAsFactors = F)\n",
    "wishedGenetypes <- wishedGenetypes$V1\n",
    "if (length(wishedGenetypes)>1) {cat(genetypeRemoveFile,\" must contain only exlcude or include list\\n\");q(\"no\")}\n",
    "if(length(grep(\"include\\\\s*:|exclude\\\\s*:\",wishedGenetypes,perl=T))!=1) {cat(genetypeRemoveFile,\"must contain word include: or exclude:\\n\"); q(\"no\")}\n",
    "if(length(grep(\"include\\\\s*:\",wishedGenetypes[1],perl=T))==1) {\n",
    "  geneTypeInclList <- gsub(\"include\\\\s*:\\\\s*(.+)\",\"\\\\1\",wishedGenetypes,perl=T)\n",
    "  geneTypeInclList <- unlist(strsplit(geneTypeInclList,\"\\\\s\"))\n",
    "  wh <- which(nchar(geneTypeInclList)>0)\n",
    "  geneTypeInclList <- geneTypeInclList[wh]\n",
    "}\n",
    "if(length(grep(\"exclude\\\\s*:\",wishedGenetypes[1],perl=T))==1) {\n",
    "  geneTypeExclList <- gsub(\"exclude\\\\s*:\\\\s*(.+)\",\"\\\\1\",wishedGenetypes,perl=T)\n",
    "  geneTypeExclList <- unlist(strsplit(geneTypeExclList,\"\\\\s\"))\n",
    "  wh <- which(nchar(geneTypeExclList)>0)\n",
    "  geneTypeExclList <- geneTypeExclList[wh]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2f109299",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 12 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>R-1090</th><th scope=col>R-1091</th><th scope=col>R-1092</th><th scope=col>R-1093</th><th scope=col>R-1094</th><th scope=col>R-1095</th><th scope=col>R-1096</th><th scope=col>R-1097</th><th scope=col>R-1098</th><th scope=col>R-1099</th><th scope=col>R-1100</th><th scope=col>R-1101</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000223972.4</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000227232.4</th><td>14</td><td>15</td><td>20</td><td>13</td><td>32</td><td>11</td><td>12</td><td>8</td><td>11</td><td>16</td><td>16</td><td>73</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000243485.2</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237613.2</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000268020.2</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000240361.1</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 12 of type int\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & R-1090 & R-1091 & R-1092 & R-1093 & R-1094 & R-1095 & R-1096 & R-1097 & R-1098 & R-1099 & R-1100 & R-1101\\\\\n",
       "\\hline\n",
       "\tENSG00000223972.4 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000227232.4 & 14 & 15 & 20 & 13 & 32 & 11 & 12 & 8 & 11 & 16 & 16 & 73\\\\\n",
       "\tENSG00000243485.2 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000237613.2 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000268020.2 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000240361.1 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 12 of type int\n",
       "\n",
       "| <!--/--> | R-1090 | R-1091 | R-1092 | R-1093 | R-1094 | R-1095 | R-1096 | R-1097 | R-1098 | R-1099 | R-1100 | R-1101 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000223972.4 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000227232.4 | 14 | 15 | 20 | 13 | 32 | 11 | 12 | 8 | 11 | 16 | 16 | 73 |\n",
       "| ENSG00000243485.2 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000237613.2 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000268020.2 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000240361.1 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "                  R-1090 R-1091 R-1092 R-1093 R-1094 R-1095 R-1096 R-1097\n",
       "ENSG00000223972.4  0      0      0      0      0      0      0     0     \n",
       "ENSG00000227232.4 14     15     20     13     32     11     12     8     \n",
       "ENSG00000243485.2  0      0      0      0      0      0      0     0     \n",
       "ENSG00000237613.2  0      0      0      0      0      0      0     0     \n",
       "ENSG00000268020.2  0      0      0      0      0      0      0     0     \n",
       "ENSG00000240361.1  0      0      0      0      0      0      0     0     \n",
       "                  R-1098 R-1099 R-1100 R-1101\n",
       "ENSG00000223972.4  0      0      0      0    \n",
       "ENSG00000227232.4 11     16     16     73    \n",
       "ENSG00000243485.2  0      0      0      0    \n",
       "ENSG00000237613.2  0      0      0      0    \n",
       "ENSG00000268020.2  0      0      0      0    \n",
       "ENSG00000240361.1  0      0      0      0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>57952</li><li>12</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 57952\n",
       "\\item 12\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 57952\n",
       "2. 12\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 57952    12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ENSG00000223972.4'</li><li>'ENSG00000227232.4'</li><li>'ENSG00000243485.2'</li><li>'ENSG00000237613.2'</li><li>'ENSG00000268020.2'</li><li>'ENSG00000240361.1'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ENSG00000223972.4'\n",
       "\\item 'ENSG00000227232.4'\n",
       "\\item 'ENSG00000243485.2'\n",
       "\\item 'ENSG00000237613.2'\n",
       "\\item 'ENSG00000268020.2'\n",
       "\\item 'ENSG00000240361.1'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ENSG00000223972.4'\n",
       "2. 'ENSG00000227232.4'\n",
       "3. 'ENSG00000243485.2'\n",
       "4. 'ENSG00000237613.2'\n",
       "5. 'ENSG00000268020.2'\n",
       "6. 'ENSG00000240361.1'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ENSG00000223972.4\" \"ENSG00000227232.4\" \"ENSG00000243485.2\"\n",
       "[4] \"ENSG00000237613.2\" \"ENSG00000268020.2\" \"ENSG00000240361.1\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "load(file = featureCountFile)\n",
    "counts <- as.matrix(fcountGenes$counts)\n",
    "head(counts)\n",
    "dim(counts)\n",
    "\n",
    "# simplify Ensg ID if necessary:\n",
    "head(rownames(counts) )\n",
    "\n",
    "rownames(counts) <- gsub(\"([^\\\\.]+)\\\\.?.*\",\"\\\\1\",rownames(counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7b9b252f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>ENSG00000223972</dt><dd>1756</dd><dt>ENSG00000227232</dt><dd>2073</dd><dt>ENSG00000243485</dt><dd>1021</dd><dt>ENSG00000237613</dt><dd>1219</dd><dt>ENSG00000268020</dt><dd>947</dd><dt>ENSG00000240361</dt><dd>940</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[ENSG00000223972] 1756\n",
       "\\item[ENSG00000227232] 2073\n",
       "\\item[ENSG00000243485] 1021\n",
       "\\item[ENSG00000237613] 1219\n",
       "\\item[ENSG00000268020] 947\n",
       "\\item[ENSG00000240361] 940\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "ENSG00000223972\n",
       ":   1756ENSG00000227232\n",
       ":   2073ENSG00000243485\n",
       ":   1021ENSG00000237613\n",
       ":   1219ENSG00000268020\n",
       ":   947ENSG00000240361\n",
       ":   940\n",
       "\n"
      ],
      "text/plain": [
       "ENSG00000223972 ENSG00000227232 ENSG00000243485 ENSG00000237613 ENSG00000268020 \n",
       "           1756            2073            1021            1219             947 \n",
       "ENSG00000240361 \n",
       "            940 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 12 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>R-1090</th><th scope=col>R-1091</th><th scope=col>R-1092</th><th scope=col>R-1093</th><th scope=col>R-1094</th><th scope=col>R-1095</th><th scope=col>R-1096</th><th scope=col>R-1097</th><th scope=col>R-1098</th><th scope=col>R-1099</th><th scope=col>R-1100</th><th scope=col>R-1101</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000223972</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000227232</th><td>14</td><td>15</td><td>20</td><td>13</td><td>32</td><td>11</td><td>12</td><td>8</td><td>11</td><td>16</td><td>16</td><td>73</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000243485</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237613</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000268020</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000240361</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 12 of type int\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & R-1090 & R-1091 & R-1092 & R-1093 & R-1094 & R-1095 & R-1096 & R-1097 & R-1098 & R-1099 & R-1100 & R-1101\\\\\n",
       "\\hline\n",
       "\tENSG00000223972 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000227232 & 14 & 15 & 20 & 13 & 32 & 11 & 12 & 8 & 11 & 16 & 16 & 73\\\\\n",
       "\tENSG00000243485 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000237613 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000268020 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000240361 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 12 of type int\n",
       "\n",
       "| <!--/--> | R-1090 | R-1091 | R-1092 | R-1093 | R-1094 | R-1095 | R-1096 | R-1097 | R-1098 | R-1099 | R-1100 | R-1101 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000223972 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000227232 | 14 | 15 | 20 | 13 | 32 | 11 | 12 | 8 | 11 | 16 | 16 | 73 |\n",
       "| ENSG00000243485 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000237613 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000268020 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000240361 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "                R-1090 R-1091 R-1092 R-1093 R-1094 R-1095 R-1096 R-1097 R-1098\n",
       "ENSG00000223972  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000227232 14     15     20     13     32     11     12     8      11    \n",
       "ENSG00000243485  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000237613  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000268020  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000240361  0      0      0      0      0      0      0     0       0    \n",
       "                R-1099 R-1100 R-1101\n",
       "ENSG00000223972  0      0      0    \n",
       "ENSG00000227232 16     16     73    \n",
       "ENSG00000243485  0      0      0    \n",
       "ENSG00000237613  0      0      0    \n",
       "ENSG00000268020  0      0      0    \n",
       "ENSG00000240361  0      0      0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>57952</li><li>12</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 57952\n",
       "\\item 12\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 57952\n",
       "2. 12\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 57952    12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "57952"
      ],
      "text/latex": [
       "57952"
      ],
      "text/markdown": [
       "57952"
      ],
      "text/plain": [
       "[1] 57952"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 4 × 4 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>R-1090</th><th scope=col>R-1091</th><th scope=col>R-1092</th><th scope=col>R-1093</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000223972</th><td>0.000000</td><td>0.00000</td><td>0.000000</td><td>0.000000</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000227232</th><td>6.753497</td><td>7.23589</td><td>9.647853</td><td>6.271105</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000243485</th><td>0.000000</td><td>0.00000</td><td>0.000000</td><td>0.000000</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237613</th><td>0.000000</td><td>0.00000</td><td>0.000000</td><td>0.000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 4 × 4 of type dbl\n",
       "\\begin{tabular}{r|llll}\n",
       "  & R-1090 & R-1091 & R-1092 & R-1093\\\\\n",
       "\\hline\n",
       "\tENSG00000223972 & 0.000000 & 0.00000 & 0.000000 & 0.000000\\\\\n",
       "\tENSG00000227232 & 6.753497 & 7.23589 & 9.647853 & 6.271105\\\\\n",
       "\tENSG00000243485 & 0.000000 & 0.00000 & 0.000000 & 0.000000\\\\\n",
       "\tENSG00000237613 & 0.000000 & 0.00000 & 0.000000 & 0.000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 4 × 4 of type dbl\n",
       "\n",
       "| <!--/--> | R-1090 | R-1091 | R-1092 | R-1093 |\n",
       "|---|---|---|---|---|\n",
       "| ENSG00000223972 | 0.000000 | 0.00000 | 0.000000 | 0.000000 |\n",
       "| ENSG00000227232 | 6.753497 | 7.23589 | 9.647853 | 6.271105 |\n",
       "| ENSG00000243485 | 0.000000 | 0.00000 | 0.000000 | 0.000000 |\n",
       "| ENSG00000237613 | 0.000000 | 0.00000 | 0.000000 | 0.000000 |\n",
       "\n"
      ],
      "text/plain": [
       "                R-1090   R-1091  R-1092   R-1093  \n",
       "ENSG00000223972 0.000000 0.00000 0.000000 0.000000\n",
       "ENSG00000227232 6.753497 7.23589 9.647853 6.271105\n",
       "ENSG00000243485 0.000000 0.00000 0.000000 0.000000\n",
       "ENSG00000237613 0.000000 0.00000 0.000000 0.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TPMs:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 4 × 4 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>R-1090</th><th scope=col>R-1091</th><th scope=col>R-1092</th><th scope=col>R-1093</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000223972</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000227232</th><td>14</td><td>15</td><td>20</td><td>13</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000243485</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237613</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 4 × 4 of type int\n",
       "\\begin{tabular}{r|llll}\n",
       "  & R-1090 & R-1091 & R-1092 & R-1093\\\\\n",
       "\\hline\n",
       "\tENSG00000223972 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000227232 & 14 & 15 & 20 & 13\\\\\n",
       "\tENSG00000243485 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000237613 &  0 &  0 &  0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 4 × 4 of type int\n",
       "\n",
       "| <!--/--> | R-1090 | R-1091 | R-1092 | R-1093 |\n",
       "|---|---|---|---|---|\n",
       "| ENSG00000223972 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000227232 | 14 | 15 | 20 | 13 |\n",
       "| ENSG00000243485 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000237613 |  0 |  0 |  0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "                R-1090 R-1091 R-1092 R-1093\n",
       "ENSG00000223972  0      0      0      0    \n",
       "ENSG00000227232 14     15     20     13    \n",
       "ENSG00000243485  0      0      0      0    \n",
       "ENSG00000237613  0      0      0      0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 4 × 4 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>R-1090</th><th scope=col>R-1091</th><th scope=col>R-1092</th><th scope=col>R-1093</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000223972</th><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.00000</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000227232</th><td>2.546357</td><td>2.929086</td><td>3.548594</td><td>3.32791</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000243485</th><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.00000</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237613</th><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.00000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 4 × 4 of type dbl\n",
       "\\begin{tabular}{r|llll}\n",
       "  & R-1090 & R-1091 & R-1092 & R-1093\\\\\n",
       "\\hline\n",
       "\tENSG00000223972 & 0.000000 & 0.000000 & 0.000000 & 0.00000\\\\\n",
       "\tENSG00000227232 & 2.546357 & 2.929086 & 3.548594 & 3.32791\\\\\n",
       "\tENSG00000243485 & 0.000000 & 0.000000 & 0.000000 & 0.00000\\\\\n",
       "\tENSG00000237613 & 0.000000 & 0.000000 & 0.000000 & 0.00000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 4 × 4 of type dbl\n",
       "\n",
       "| <!--/--> | R-1090 | R-1091 | R-1092 | R-1093 |\n",
       "|---|---|---|---|---|\n",
       "| ENSG00000223972 | 0.000000 | 0.000000 | 0.000000 | 0.00000 |\n",
       "| ENSG00000227232 | 2.546357 | 2.929086 | 3.548594 | 3.32791 |\n",
       "| ENSG00000243485 | 0.000000 | 0.000000 | 0.000000 | 0.00000 |\n",
       "| ENSG00000237613 | 0.000000 | 0.000000 | 0.000000 | 0.00000 |\n",
       "\n"
      ],
      "text/plain": [
       "                R-1090   R-1091   R-1092   R-1093 \n",
       "ENSG00000223972 0.000000 0.000000 0.000000 0.00000\n",
       "ENSG00000227232 2.546357 2.929086 3.548594 3.32791\n",
       "ENSG00000243485 0.000000 0.000000 0.000000 0.00000\n",
       "ENSG00000237613 0.000000 0.000000 0.000000 0.00000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# first normalize by gene length, then by sum of normalized expr per sample\n",
    "geneLengths <- fcountGenes$annotation[,\"Length\"]\n",
    "names(geneLengths) <- fcountGenes$annotation[,\"GeneID\"]\n",
    "names(geneLengths) <- gsub(\"(\\\\w+)\\\\.?\\\\d*\",\"\\\\1\",names(geneLengths))\n",
    "head(geneLengths)\n",
    "\n",
    "TPMs <- counts\n",
    "head(TPMs)\n",
    "dim(TPMs[,names(totalReads)])\n",
    "length(geneLengths)\n",
    "TPMs[,names(totalReads)] <- sapply(names(totalReads),function(x) TPMs[,x]/(geneLengths/10^3) )\n",
    "\n",
    "TPMs[1:4,1:4]\n",
    "TPMsums <- colSums(TPMs)\n",
    "TPMs[,names(totalReads)] <- sapply(names(TPMsums),function(x) TPMs[,x]/(TPMsums[x]/10^6))\n",
    "cat(\"TPMs:\\n\")\n",
    "counts[1:4,1:4]\n",
    "TPMs[1:4,1:4]\n",
    "\n",
    "if (extraEnsgIDFile!=\"n\") {\n",
    "  extraEnsgID <- readLines(extraEnsgIDFile)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "08e6d292",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>cell_line</th><th scope=col>group</th><th scope=col>name</th><th scope=col>group_full</th><th scope=col>RNA_concentation</th><th scope=col>ratio_260_280</th><th scope=col>ratio_260_230</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>R-1090</td><td>UWB1.289</td><td>0    </td><td>R-1090_ctr  </td><td>UWB1.289_ctr  </td><td>3804.7</td><td>1.96</td><td>2.36</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>R-1091</td><td>UWB1.289</td><td>0    </td><td>R-1091_ctr  </td><td>UWB1.289_ctr  </td><td>3383.5</td><td>1.98</td><td>2.34</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>R-1092</td><td>UWB1.289</td><td>CT913</td><td>R-1092_CT913</td><td>UWB1.289_CT913</td><td>2859.8</td><td>1.95</td><td>2.24</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>R-1093</td><td>UWB1.289</td><td>CT913</td><td>R-1093_CT913</td><td>UWB1.289_CT913</td><td>3208.5</td><td>1.94</td><td>2.13</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>R-1094</td><td>UWB1.289</td><td>CT_M1</td><td>R-1094_CT_M1</td><td>UWB1.289_CT_M1</td><td>2805.7</td><td>1.95</td><td>2.11</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>R-1095</td><td>UWB1.289</td><td>CT_M1</td><td>R-1095_CT_M1</td><td>UWB1.289_CT_M1</td><td>2958.1</td><td>1.94</td><td>2.28</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & ID & cell\\_line & group & name & group\\_full & RNA\\_concentation & ratio\\_260\\_280 & ratio\\_260\\_230\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & R-1090 & UWB1.289 & 0     & R-1090\\_ctr   & UWB1.289\\_ctr   & 3804.7 & 1.96 & 2.36\\\\\n",
       "\t2 & R-1091 & UWB1.289 & 0     & R-1091\\_ctr   & UWB1.289\\_ctr   & 3383.5 & 1.98 & 2.34\\\\\n",
       "\t3 & R-1092 & UWB1.289 & CT913 & R-1092\\_CT913 & UWB1.289\\_CT913 & 2859.8 & 1.95 & 2.24\\\\\n",
       "\t4 & R-1093 & UWB1.289 & CT913 & R-1093\\_CT913 & UWB1.289\\_CT913 & 3208.5 & 1.94 & 2.13\\\\\n",
       "\t5 & R-1094 & UWB1.289 & CT\\_M1 & R-1094\\_CT\\_M1 & UWB1.289\\_CT\\_M1 & 2805.7 & 1.95 & 2.11\\\\\n",
       "\t6 & R-1095 & UWB1.289 & CT\\_M1 & R-1095\\_CT\\_M1 & UWB1.289\\_CT\\_M1 & 2958.1 & 1.94 & 2.28\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 8\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | cell_line &lt;chr&gt; | group &lt;chr&gt; | name &lt;chr&gt; | group_full &lt;chr&gt; | RNA_concentation &lt;chr&gt; | ratio_260_280 &lt;chr&gt; | ratio_260_230 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | R-1090 | UWB1.289 | 0     | R-1090_ctr   | UWB1.289_ctr   | 3804.7 | 1.96 | 2.36 |\n",
       "| 2 | R-1091 | UWB1.289 | 0     | R-1091_ctr   | UWB1.289_ctr   | 3383.5 | 1.98 | 2.34 |\n",
       "| 3 | R-1092 | UWB1.289 | CT913 | R-1092_CT913 | UWB1.289_CT913 | 2859.8 | 1.95 | 2.24 |\n",
       "| 4 | R-1093 | UWB1.289 | CT913 | R-1093_CT913 | UWB1.289_CT913 | 3208.5 | 1.94 | 2.13 |\n",
       "| 5 | R-1094 | UWB1.289 | CT_M1 | R-1094_CT_M1 | UWB1.289_CT_M1 | 2805.7 | 1.95 | 2.11 |\n",
       "| 6 | R-1095 | UWB1.289 | CT_M1 | R-1095_CT_M1 | UWB1.289_CT_M1 | 2958.1 | 1.94 | 2.28 |\n",
       "\n"
      ],
      "text/plain": [
       "  ID     cell_line group name         group_full     RNA_concentation\n",
       "1 R-1090 UWB1.289  0     R-1090_ctr   UWB1.289_ctr   3804.7          \n",
       "2 R-1091 UWB1.289  0     R-1091_ctr   UWB1.289_ctr   3383.5          \n",
       "3 R-1092 UWB1.289  CT913 R-1092_CT913 UWB1.289_CT913 2859.8          \n",
       "4 R-1093 UWB1.289  CT913 R-1093_CT913 UWB1.289_CT913 3208.5          \n",
       "5 R-1094 UWB1.289  CT_M1 R-1094_CT_M1 UWB1.289_CT_M1 2805.7          \n",
       "6 R-1095 UWB1.289  CT_M1 R-1095_CT_M1 UWB1.289_CT_M1 2958.1          \n",
       "  ratio_260_280 ratio_260_230\n",
       "1 1.96          2.36         \n",
       "2 1.98          2.34         \n",
       "3 1.95          2.24         \n",
       "4 1.94          2.13         \n",
       "5 1.95          2.11         \n",
       "6 1.94          2.28         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load sample annotations (cell culture, RNA concentration, extraction date, ...)\n",
    "sampleAnnot <- read.csv(file = sampleAnnotFile,sep = \"\\t\", dec=\",\", stringsAsFactors = F )\n",
    "head(sampleAnnot)\n",
    "wh <- which(sampleAnnot$ID==\"\")\n",
    "if(length(wh)>0) {sampleAnnot <- sampleAnnot[-wh,]}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7e525fed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GeneID</th><th scope=col>Chr</th><th scope=col>Start</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000223972.4</td><td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                   </td><td>11869;11872;11874;12010;12179;12595;12613;12613;12613;12975;13221;13221;13225;13403;13453;13661                                                                                                                                                                                                                                          </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000227232.4</td><td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1</td><td>14363;14363;14363;14404;14411;14970;14970;14970;15000;15005;15796;15796;15796;15796;15904;15904;16607;16607;16607;16607;16748;16854;16854;16858;16858;16858;17233;17233;17233;17233;17233;17498;17602;17602;17606;17606;17606;17915;17915;17915;17915;17915;18268;18268;18268;18268;24734;24737;24738;24738;24738;29321;29321;29534;29534</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENSG00000243485.2</td><td>chr1;chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                                                                     </td><td>29554;30267;30366;30564;30976;30976                                                                                                                                                                                                                                                                                                      </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENSG00000237613.2</td><td>chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                                                                          </td><td>34554;35245;35277;35721;35721                                                                                                                                                                                                                                                                                                            </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENSG00000268020.2</td><td>chr1;chr1;chr1                                                                                                                                                                                                                                                                    </td><td>52473;53049;54830                                                                                                                                                                                                                                                                                                                        </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENSG00000240361.1</td><td>chr1                                                                                                                                                                                                                                                                              </td><td>62948                                                                                                                                                                                                                                                                                                                                    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & GeneID & Chr & Start\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000223972.4 & chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                    & 11869;11872;11874;12010;12179;12595;12613;12613;12613;12975;13221;13221;13225;13403;13453;13661                                                                                                                                                                                                                                          \\\\\n",
       "\t2 & ENSG00000227232.4 & chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1 & 14363;14363;14363;14404;14411;14970;14970;14970;15000;15005;15796;15796;15796;15796;15904;15904;16607;16607;16607;16607;16748;16854;16854;16858;16858;16858;17233;17233;17233;17233;17233;17498;17602;17602;17606;17606;17606;17915;17915;17915;17915;17915;18268;18268;18268;18268;24734;24737;24738;24738;24738;29321;29321;29534;29534\\\\\n",
       "\t3 & ENSG00000243485.2 & chr1;chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                                                                      & 29554;30267;30366;30564;30976;30976                                                                                                                                                                                                                                                                                                      \\\\\n",
       "\t4 & ENSG00000237613.2 & chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                                                                           & 34554;35245;35277;35721;35721                                                                                                                                                                                                                                                                                                            \\\\\n",
       "\t5 & ENSG00000268020.2 & chr1;chr1;chr1                                                                                                                                                                                                                                                                     & 52473;53049;54830                                                                                                                                                                                                                                                                                                                        \\\\\n",
       "\t6 & ENSG00000240361.1 & chr1                                                                                                                                                                                                                                                                               & 62948                                                                                                                                                                                                                                                                                                                                    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | GeneID &lt;chr&gt; | Chr &lt;chr&gt; | Start &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | ENSG00000223972.4 | chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                    | 11869;11872;11874;12010;12179;12595;12613;12613;12613;12975;13221;13221;13225;13403;13453;13661                                                                                                                                                                                                                                           |\n",
       "| 2 | ENSG00000227232.4 | chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1 | 14363;14363;14363;14404;14411;14970;14970;14970;15000;15005;15796;15796;15796;15796;15904;15904;16607;16607;16607;16607;16748;16854;16854;16858;16858;16858;17233;17233;17233;17233;17233;17498;17602;17602;17606;17606;17606;17915;17915;17915;17915;17915;18268;18268;18268;18268;24734;24737;24738;24738;24738;29321;29321;29534;29534 |\n",
       "| 3 | ENSG00000243485.2 | chr1;chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                                                                      | 29554;30267;30366;30564;30976;30976                                                                                                                                                                                                                                                                                                       |\n",
       "| 4 | ENSG00000237613.2 | chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                                                                           | 34554;35245;35277;35721;35721                                                                                                                                                                                                                                                                                                             |\n",
       "| 5 | ENSG00000268020.2 | chr1;chr1;chr1                                                                                                                                                                                                                                                                     | 52473;53049;54830                                                                                                                                                                                                                                                                                                                         |\n",
       "| 6 | ENSG00000240361.1 | chr1                                                                                                                                                                                                                                                                               | 62948                                                                                                                                                                                                                                                                                                                                     |\n",
       "\n"
      ],
      "text/plain": [
       "  GeneID           \n",
       "1 ENSG00000223972.4\n",
       "2 ENSG00000227232.4\n",
       "3 ENSG00000243485.2\n",
       "4 ENSG00000237613.2\n",
       "5 ENSG00000268020.2\n",
       "6 ENSG00000240361.1\n",
       "  Chr                                                                                                                                                                                                                                                                               \n",
       "1 chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                   \n",
       "2 chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1\n",
       "3 chr1;chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                                                                     \n",
       "4 chr1;chr1;chr1;chr1;chr1                                                                                                                                                                                                                                                          \n",
       "5 chr1;chr1;chr1                                                                                                                                                                                                                                                                    \n",
       "6 chr1                                                                                                                                                                                                                                                                              \n",
       "  Start                                                                                                                                                                                                                                                                                                                                    \n",
       "1 11869;11872;11874;12010;12179;12595;12613;12613;12613;12975;13221;13221;13225;13403;13453;13661                                                                                                                                                                                                                                          \n",
       "2 14363;14363;14363;14404;14411;14970;14970;14970;15000;15005;15796;15796;15796;15796;15904;15904;16607;16607;16607;16607;16748;16854;16854;16858;16858;16858;17233;17233;17233;17233;17233;17498;17602;17602;17606;17606;17606;17915;17915;17915;17915;17915;18268;18268;18268;18268;24734;24737;24738;24738;24738;29321;29321;29534;29534\n",
       "3 29554;30267;30366;30564;30976;30976                                                                                                                                                                                                                                                                                                      \n",
       "4 34554;35245;35277;35721;35721                                                                                                                                                                                                                                                                                                            \n",
       "5 52473;53049;54830                                                                                                                                                                                                                                                                                                                        \n",
       "6 62948                                                                                                                                                                                                                                                                                                                                    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if(!all(colnames(counts) %in% sampleAnnot$ID)) {\n",
    "  cat(paste(\"feature count names:\\n\", paste(colnames(counts),collapse=\", \"),\n",
    "            \"\\nnot found in sampleAnnot ID:\\n\",paste(sampleAnnot$ID,collapse=\", \"),\"\\n\"))\n",
    "  #q(\"no\")\n",
    "  \n",
    "  commonSamples <- intersect(colnames(counts), sampleAnnot$ID)\n",
    "  print(paste(length(commonSamples), \"samples appear in both, featureCount (\",length(colnames(counts)),\n",
    "              \") and sampleAnnot (\",length(sampleAnnot$ID),\")\"))\n",
    "  print(paste(\"will use only common samples...\"))\n",
    "  counts <- counts[,commonSamples]\n",
    "  sampleAnnot <- sampleAnnot[sampleAnnot$ID %in% commonSamples,]\n",
    "}\n",
    "rownames(sampleAnnot) <- sampleAnnot$ID\n",
    "\n",
    "# gene positions\n",
    "genePos <- fcountGenes$annotation[c(\"GeneID\",\"Chr\",\"Start\")]\n",
    "head(genePos)\n",
    "genePos$GeneID<- gsub(\"(\\\\w+)\\\\.?\\\\d*\",\"\\\\1\",genePos$GeneID)\n",
    "genePos$Chr<- gsub(\"(\\\\w+).*\",\"\\\\1\",genePos$Chr)\n",
    "genePos$Start<- gsub(\"(\\\\w+).*\",\"\\\\1\",genePos$Start)\n",
    "rownames(genePos) <- make.unique(genePos$GeneID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a94322e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 12 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>R-1090</th><th scope=col>R-1091</th><th scope=col>R-1092</th><th scope=col>R-1093</th><th scope=col>R-1094</th><th scope=col>R-1095</th><th scope=col>R-1096</th><th scope=col>R-1097</th><th scope=col>R-1098</th><th scope=col>R-1099</th><th scope=col>R-1100</th><th scope=col>R-1101</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000223972</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000227232</th><td>14</td><td>15</td><td>20</td><td>13</td><td>32</td><td>11</td><td>12</td><td>8</td><td>11</td><td>16</td><td>16</td><td>73</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000243485</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237613</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000268020</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000240361</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 12 of type int\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & R-1090 & R-1091 & R-1092 & R-1093 & R-1094 & R-1095 & R-1096 & R-1097 & R-1098 & R-1099 & R-1100 & R-1101\\\\\n",
       "\\hline\n",
       "\tENSG00000223972 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000227232 & 14 & 15 & 20 & 13 & 32 & 11 & 12 & 8 & 11 & 16 & 16 & 73\\\\\n",
       "\tENSG00000243485 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000237613 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000268020 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000240361 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 12 of type int\n",
       "\n",
       "| <!--/--> | R-1090 | R-1091 | R-1092 | R-1093 | R-1094 | R-1095 | R-1096 | R-1097 | R-1098 | R-1099 | R-1100 | R-1101 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000223972 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000227232 | 14 | 15 | 20 | 13 | 32 | 11 | 12 | 8 | 11 | 16 | 16 | 73 |\n",
       "| ENSG00000243485 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000237613 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000268020 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000240361 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "                R-1090 R-1091 R-1092 R-1093 R-1094 R-1095 R-1096 R-1097 R-1098\n",
       "ENSG00000223972  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000227232 14     15     20     13     32     11     12     8      11    \n",
       "ENSG00000243485  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000237613  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000268020  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000240361  0      0      0      0      0      0      0     0       0    \n",
       "                R-1099 R-1100 R-1101\n",
       "ENSG00000223972  0      0      0    \n",
       "ENSG00000227232 16     16     73    \n",
       "ENSG00000243485  0      0      0    \n",
       "ENSG00000237613  0      0      0    \n",
       "ENSG00000268020  0      0      0    \n",
       "ENSG00000240361  0      0      0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"removing 528 genes from RPKM\"\n",
      "[1] \"528 genes to remove ( rRNA, Mt_rRNA )\"\n"
     ]
    }
   ],
   "source": [
    "# remove unwanted gene types:\n",
    "head(counts)\n",
    "# include:\n",
    "if (exists(\"geneTypeInclList\")) {\n",
    "  geneIds2incl <- annot$ensgID[which(annot$geneType %in% geneTypeInclList)]\n",
    "  head(geneIds2incl)\n",
    "  wh2incl <- which(rownames(counts) %in% geneIds2incl)\n",
    "  print(paste(\"including\",length(intersect(rownames(counts),geneIds2incl)),\"genes from RPKM\"))\n",
    "  if(length(wh2incl)>0) {\n",
    "    print(paste(length(intersect(rownames(counts),geneIds2incl)),\"genes to keep (\",paste(geneTypeInclList,collapse=\", \"),\")\"))\n",
    "    counts <- counts[wh2incl,]\n",
    "  }\n",
    "}\n",
    "# exclude:\n",
    "if (exists(\"geneTypeExclList\")) {\n",
    "  geneIds2remove <- annot$ensgID[which(annot$geneType %in% geneTypeExclList)]\n",
    "  head(geneIds2remove)\n",
    "  wh2rm <- which(rownames(counts) %in% geneIds2remove)\n",
    "  length(rownames(counts))\n",
    "  print(paste(\"removing\",length(intersect(rownames(counts),geneIds2remove)),\"genes from RPKM\"))\n",
    "  if(length(wh2rm)>0) {\n",
    "    print(paste(length(intersect(rownames(counts),geneIds2remove)),\"genes to remove (\",paste(geneTypeExclList,collapse=\", \"),\")\"))\n",
    "    counts <- counts[-wh2rm,]\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "40a86392",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>57424</li><li>12</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 57424\n",
       "\\item 12\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 57424\n",
       "2. 12\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 57424    12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "colnames counts: R-1090; R-1091; R-1092; R-1093; R-1094; R-1095; R-1096; R-1097; R-1098; R-1099; R-1100; R-1101 \n"
     ]
    }
   ],
   "source": [
    "dim(counts)\n",
    "cat(\"colnames counts:\",paste(colnames(counts),collapse=\"; \"),\"\\n\")\n",
    "# only keep samples in annot where there is count data:\n",
    "sampleAnnot <- sampleAnnot[colnames(counts),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6d10939a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load deseq comparisons:\n",
    "deseqCompareLines <- read.csv(deseqCompareFile,header=F,stringsAsFactors = F,sep=\"?\")\n",
    "deseqCompareLines <- deseqCompareLines$V1\n",
    "deseqComparisons <- NULL\n",
    "l <- deseqCompareLines[1]\n",
    "for (l in deseqCompareLines) {\n",
    "  var <- gsub(\"(\\\\w+).+\",\"\\\\1\",l)\n",
    "  comps <- unlist(strsplit(gsub(\"\\\\w+\\\\s*:\\\\s*(.+)\",\"\\\\1\",l),\",\"))\n",
    "  c <- comps[1]\n",
    "  tmp <- NULL\n",
    "  for (c in comps) {\n",
    "    c1 <- gsub(\" *([^ \\t]+) +.+\",\"\\\\1\",c); c2 <- gsub(\".+ +([^ \\t]+) *$\",\"\\\\1\",c);\n",
    "    c2 <- gsub(\"[ \\t]$\",\"\",c2); c1 <- gsub(\"[ \\t]$\",\"\",c1) # remove trailing space if necessary\n",
    "    c2 <- gsub(\"^[ \\t]\",\"\",c2); c1 <- gsub(\"^[ \\t]\",\"\",c1) # remove leading space if necessary\n",
    "    tmp <- cbind(tmp ,c(c1,c2))\n",
    "  }\n",
    "  deseqComparisons <- c(deseqComparisons,list(tmp))\n",
    "  names(deseqComparisons)[length(deseqComparisons)] <- var\n",
    "}\n",
    "\n",
    "var <- names(deseqComparisons)[1]\n",
    "for (var in names(deseqComparisons)){ \n",
    "  if(!var %in% colnames(sampleAnnot)) {\n",
    "    cat(\"'\",var,\"'from deseq comparisons not a variable in sample annotation\\n\");\n",
    "  }\n",
    "  vals<-unique(as.character(deseqComparisons[var][[1]]))\n",
    "  v <- vals[1]\n",
    "  for (v in vals) { \n",
    "    if(!v %in% sampleAnnot[,var]) {\n",
    "      cat(\"value '\",v,\"' from deseq comparison of\",var,\" not a value in sample annotation, or no counts / no bam file\\n\")\n",
    "      q(\"no\")\n",
    "    }\n",
    "  }\n",
    "}\n",
    "# convert to factors:\n",
    "for(i in 2:ncol(sampleAnnot)) { sampleAnnot[,i] <- as.factor(sampleAnnot[,i])}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bb34da3",
   "metadata": {},
   "source": [
    "## DESeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bf5a8fb3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 12 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>R-1090</th><th scope=col>R-1091</th><th scope=col>R-1092</th><th scope=col>R-1093</th><th scope=col>R-1094</th><th scope=col>R-1095</th><th scope=col>R-1096</th><th scope=col>R-1097</th><th scope=col>R-1098</th><th scope=col>R-1099</th><th scope=col>R-1100</th><th scope=col>R-1101</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000223972</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000227232</th><td>14</td><td>15</td><td>20</td><td>13</td><td>32</td><td>11</td><td>12</td><td>8</td><td>11</td><td>16</td><td>16</td><td>73</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000243485</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237613</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000268020</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000240361</th><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 12 of type int\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & R-1090 & R-1091 & R-1092 & R-1093 & R-1094 & R-1095 & R-1096 & R-1097 & R-1098 & R-1099 & R-1100 & R-1101\\\\\n",
       "\\hline\n",
       "\tENSG00000223972 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000227232 & 14 & 15 & 20 & 13 & 32 & 11 & 12 & 8 & 11 & 16 & 16 & 73\\\\\n",
       "\tENSG00000243485 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000237613 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000268020 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\tENSG00000240361 &  0 &  0 &  0 &  0 &  0 &  0 &  0 & 0 &  0 &  0 &  0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 12 of type int\n",
       "\n",
       "| <!--/--> | R-1090 | R-1091 | R-1092 | R-1093 | R-1094 | R-1095 | R-1096 | R-1097 | R-1098 | R-1099 | R-1100 | R-1101 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000223972 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000227232 | 14 | 15 | 20 | 13 | 32 | 11 | 12 | 8 | 11 | 16 | 16 | 73 |\n",
       "| ENSG00000243485 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000237613 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000268020 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "| ENSG00000240361 |  0 |  0 |  0 |  0 |  0 |  0 |  0 | 0 |  0 |  0 |  0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "                R-1090 R-1091 R-1092 R-1093 R-1094 R-1095 R-1096 R-1097 R-1098\n",
       "ENSG00000223972  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000227232 14     15     20     13     32     11     12     8      11    \n",
       "ENSG00000243485  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000237613  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000268020  0      0      0      0      0      0      0     0       0    \n",
       "ENSG00000240361  0      0      0      0      0      0      0     0       0    \n",
       "                R-1099 R-1100 R-1101\n",
       "ENSG00000223972  0      0      0    \n",
       "ENSG00000227232 16     16     73    \n",
       "ENSG00000243485  0      0      0    \n",
       "ENSG00000237613  0      0      0    \n",
       "ENSG00000268020  0      0      0    \n",
       "ENSG00000240361  0      0      0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 12 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>cell_line</th><th scope=col>group</th><th scope=col>name</th><th scope=col>group_full</th><th scope=col>RNA_concentation</th><th scope=col>ratio_260_280</th><th scope=col>ratio_260_230</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>R-1090</th><td>R-1090</td><td>UWB1.289      </td><td>0    </td><td>R-1090_ctr        </td><td>UWB1.289_ctr        </td><td>3804.7</td><td>1.96</td><td>2.36</td></tr>\n",
       "\t<tr><th scope=row>R-1091</th><td>R-1091</td><td>UWB1.289      </td><td>0    </td><td>R-1091_ctr        </td><td>UWB1.289_ctr        </td><td>3383.5</td><td>1.98</td><td>2.34</td></tr>\n",
       "\t<tr><th scope=row>R-1092</th><td>R-1092</td><td>UWB1.289      </td><td>CT913</td><td>R-1092_CT913      </td><td>UWB1.289_CT913      </td><td>2859.8</td><td>1.95</td><td>2.24</td></tr>\n",
       "\t<tr><th scope=row>R-1093</th><td>R-1093</td><td>UWB1.289      </td><td>CT913</td><td>R-1093_CT913      </td><td>UWB1.289_CT913      </td><td>3208.5</td><td>1.94</td><td>2.13</td></tr>\n",
       "\t<tr><th scope=row>R-1094</th><td>R-1094</td><td>UWB1.289      </td><td>CT_M1</td><td>R-1094_CT_M1      </td><td>UWB1.289_CT_M1      </td><td>2805.7</td><td>1.95</td><td>2.11</td></tr>\n",
       "\t<tr><th scope=row>R-1095</th><td>R-1095</td><td>UWB1.289      </td><td>CT_M1</td><td>R-1095_CT_M1      </td><td>UWB1.289_CT_M1      </td><td>2958.1</td><td>1.94</td><td>2.28</td></tr>\n",
       "\t<tr><th scope=row>R-1096</th><td>R-1096</td><td>UWB1.289+BRCA1</td><td>0    </td><td>R-1096+BRCA1_ctr  </td><td>UWB1.289_BRCA1_ctr  </td><td>1601.6</td><td>1.9 </td><td>2.29</td></tr>\n",
       "\t<tr><th scope=row>R-1097</th><td>R-1097</td><td>UWB1.289+BRCA1</td><td>0    </td><td>R-1097+BRCA1_ctr  </td><td>UWB1.289_BRCA1_ctr  </td><td>1406.1</td><td>1.74</td><td>2.04</td></tr>\n",
       "\t<tr><th scope=row>R-1098</th><td>R-1098</td><td>UWB1.289+BRCA1</td><td>CT913</td><td>R-1098+BRCA1_CT913</td><td>UWB1.289_BRCA1_CT913</td><td>1035.5</td><td>1.88</td><td>2.23</td></tr>\n",
       "\t<tr><th scope=row>R-1099</th><td>R-1099</td><td>UWB1.289+BRCA1</td><td>CT913</td><td>R-1099+BRCA1_CT913</td><td>UWB1.289_BRCA1_CT913</td><td>1211.2</td><td>1.87</td><td>2.32</td></tr>\n",
       "\t<tr><th scope=row>R-1100</th><td>R-1100</td><td>UWB1.289+BRCA1</td><td>CT_M1</td><td>R-1100+BRCA1_CT_M1</td><td>UWB1.289_BRCA1_CT_M1</td><td>1180.9</td><td>1.87</td><td>2.26</td></tr>\n",
       "\t<tr><th scope=row>R-1101</th><td>R-1101</td><td>UWB1.289+BRCA1</td><td>CT_M1</td><td>R-1101+BRCA1_CT_M1</td><td>UWB1.289_BRCA1_CT_M1</td><td>1114.2</td><td>1.94</td><td>1.56</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 12 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & ID & cell\\_line & group & name & group\\_full & RNA\\_concentation & ratio\\_260\\_280 & ratio\\_260\\_230\\\\\n",
       "  & <chr> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\tR-1090 & R-1090 & UWB1.289       & 0     & R-1090\\_ctr         & UWB1.289\\_ctr         & 3804.7 & 1.96 & 2.36\\\\\n",
       "\tR-1091 & R-1091 & UWB1.289       & 0     & R-1091\\_ctr         & UWB1.289\\_ctr         & 3383.5 & 1.98 & 2.34\\\\\n",
       "\tR-1092 & R-1092 & UWB1.289       & CT913 & R-1092\\_CT913       & UWB1.289\\_CT913       & 2859.8 & 1.95 & 2.24\\\\\n",
       "\tR-1093 & R-1093 & UWB1.289       & CT913 & R-1093\\_CT913       & UWB1.289\\_CT913       & 3208.5 & 1.94 & 2.13\\\\\n",
       "\tR-1094 & R-1094 & UWB1.289       & CT\\_M1 & R-1094\\_CT\\_M1       & UWB1.289\\_CT\\_M1       & 2805.7 & 1.95 & 2.11\\\\\n",
       "\tR-1095 & R-1095 & UWB1.289       & CT\\_M1 & R-1095\\_CT\\_M1       & UWB1.289\\_CT\\_M1       & 2958.1 & 1.94 & 2.28\\\\\n",
       "\tR-1096 & R-1096 & UWB1.289+BRCA1 & 0     & R-1096+BRCA1\\_ctr   & UWB1.289\\_BRCA1\\_ctr   & 1601.6 & 1.9  & 2.29\\\\\n",
       "\tR-1097 & R-1097 & UWB1.289+BRCA1 & 0     & R-1097+BRCA1\\_ctr   & UWB1.289\\_BRCA1\\_ctr   & 1406.1 & 1.74 & 2.04\\\\\n",
       "\tR-1098 & R-1098 & UWB1.289+BRCA1 & CT913 & R-1098+BRCA1\\_CT913 & UWB1.289\\_BRCA1\\_CT913 & 1035.5 & 1.88 & 2.23\\\\\n",
       "\tR-1099 & R-1099 & UWB1.289+BRCA1 & CT913 & R-1099+BRCA1\\_CT913 & UWB1.289\\_BRCA1\\_CT913 & 1211.2 & 1.87 & 2.32\\\\\n",
       "\tR-1100 & R-1100 & UWB1.289+BRCA1 & CT\\_M1 & R-1100+BRCA1\\_CT\\_M1 & UWB1.289\\_BRCA1\\_CT\\_M1 & 1180.9 & 1.87 & 2.26\\\\\n",
       "\tR-1101 & R-1101 & UWB1.289+BRCA1 & CT\\_M1 & R-1101+BRCA1\\_CT\\_M1 & UWB1.289\\_BRCA1\\_CT\\_M1 & 1114.2 & 1.94 & 1.56\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 12 × 8\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | cell_line &lt;fct&gt; | group &lt;fct&gt; | name &lt;fct&gt; | group_full &lt;fct&gt; | RNA_concentation &lt;fct&gt; | ratio_260_280 &lt;fct&gt; | ratio_260_230 &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| R-1090 | R-1090 | UWB1.289       | 0     | R-1090_ctr         | UWB1.289_ctr         | 3804.7 | 1.96 | 2.36 |\n",
       "| R-1091 | R-1091 | UWB1.289       | 0     | R-1091_ctr         | UWB1.289_ctr         | 3383.5 | 1.98 | 2.34 |\n",
       "| R-1092 | R-1092 | UWB1.289       | CT913 | R-1092_CT913       | UWB1.289_CT913       | 2859.8 | 1.95 | 2.24 |\n",
       "| R-1093 | R-1093 | UWB1.289       | CT913 | R-1093_CT913       | UWB1.289_CT913       | 3208.5 | 1.94 | 2.13 |\n",
       "| R-1094 | R-1094 | UWB1.289       | CT_M1 | R-1094_CT_M1       | UWB1.289_CT_M1       | 2805.7 | 1.95 | 2.11 |\n",
       "| R-1095 | R-1095 | UWB1.289       | CT_M1 | R-1095_CT_M1       | UWB1.289_CT_M1       | 2958.1 | 1.94 | 2.28 |\n",
       "| R-1096 | R-1096 | UWB1.289+BRCA1 | 0     | R-1096+BRCA1_ctr   | UWB1.289_BRCA1_ctr   | 1601.6 | 1.9  | 2.29 |\n",
       "| R-1097 | R-1097 | UWB1.289+BRCA1 | 0     | R-1097+BRCA1_ctr   | UWB1.289_BRCA1_ctr   | 1406.1 | 1.74 | 2.04 |\n",
       "| R-1098 | R-1098 | UWB1.289+BRCA1 | CT913 | R-1098+BRCA1_CT913 | UWB1.289_BRCA1_CT913 | 1035.5 | 1.88 | 2.23 |\n",
       "| R-1099 | R-1099 | UWB1.289+BRCA1 | CT913 | R-1099+BRCA1_CT913 | UWB1.289_BRCA1_CT913 | 1211.2 | 1.87 | 2.32 |\n",
       "| R-1100 | R-1100 | UWB1.289+BRCA1 | CT_M1 | R-1100+BRCA1_CT_M1 | UWB1.289_BRCA1_CT_M1 | 1180.9 | 1.87 | 2.26 |\n",
       "| R-1101 | R-1101 | UWB1.289+BRCA1 | CT_M1 | R-1101+BRCA1_CT_M1 | UWB1.289_BRCA1_CT_M1 | 1114.2 | 1.94 | 1.56 |\n",
       "\n"
      ],
      "text/plain": [
       "       ID     cell_line      group name               group_full          \n",
       "R-1090 R-1090 UWB1.289       0     R-1090_ctr         UWB1.289_ctr        \n",
       "R-1091 R-1091 UWB1.289       0     R-1091_ctr         UWB1.289_ctr        \n",
       "R-1092 R-1092 UWB1.289       CT913 R-1092_CT913       UWB1.289_CT913      \n",
       "R-1093 R-1093 UWB1.289       CT913 R-1093_CT913       UWB1.289_CT913      \n",
       "R-1094 R-1094 UWB1.289       CT_M1 R-1094_CT_M1       UWB1.289_CT_M1      \n",
       "R-1095 R-1095 UWB1.289       CT_M1 R-1095_CT_M1       UWB1.289_CT_M1      \n",
       "R-1096 R-1096 UWB1.289+BRCA1 0     R-1096+BRCA1_ctr   UWB1.289_BRCA1_ctr  \n",
       "R-1097 R-1097 UWB1.289+BRCA1 0     R-1097+BRCA1_ctr   UWB1.289_BRCA1_ctr  \n",
       "R-1098 R-1098 UWB1.289+BRCA1 CT913 R-1098+BRCA1_CT913 UWB1.289_BRCA1_CT913\n",
       "R-1099 R-1099 UWB1.289+BRCA1 CT913 R-1099+BRCA1_CT913 UWB1.289_BRCA1_CT913\n",
       "R-1100 R-1100 UWB1.289+BRCA1 CT_M1 R-1100+BRCA1_CT_M1 UWB1.289_BRCA1_CT_M1\n",
       "R-1101 R-1101 UWB1.289+BRCA1 CT_M1 R-1101+BRCA1_CT_M1 UWB1.289_BRCA1_CT_M1\n",
       "       RNA_concentation ratio_260_280 ratio_260_230\n",
       "R-1090 3804.7           1.96          2.36         \n",
       "R-1091 3383.5           1.98          2.34         \n",
       "R-1092 2859.8           1.95          2.24         \n",
       "R-1093 3208.5           1.94          2.13         \n",
       "R-1094 2805.7           1.95          2.11         \n",
       "R-1095 2958.1           1.94          2.28         \n",
       "R-1096 1601.6           1.9           2.29         \n",
       "R-1097 1406.1           1.74          2.04         \n",
       "R-1098 1035.5           1.88          2.23         \n",
       "R-1099 1211.2           1.87          2.32         \n",
       "R-1100 1180.9           1.87          2.26         \n",
       "R-1101 1114.2           1.94          1.56         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 of 2  -  group_full:UWB1.289_ctr vs UWB1.289_CT913 \n",
      "dim of reduced dds 18112 4 \n",
      "18112 genes with at least 10 reads in sum\n",
      "18112 genes with at least 0 reads per sample\n",
      "18112 genes with at least 0 reads for min 2 sample\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stop1\n",
      "stop2\n",
      "saving all genes to csv...\n",
      "saving to  diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT913-all-genes.xls and diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT913-all-genes.csv \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in write.table(x = deseqRes, file = resTableCSVFile, quote = F, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqResTPM, file = resTableCSVFileTPM, quote = F, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqResTPMrlog, file = resTableCSVFileRlogTPM, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqRes_signif, file = resTableCSVFile, quote = F, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqResTPM_signif, file = resTableCSVFileTPM, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqResTPMrlog_signif, file = resTableCSVFileRlogTPM, :\n",
      "“appending column names to file”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 of 2  -  group_full:UWB1.289_ctr vs UWB1.289_CT_M1 \n",
      "dim of reduced dds 18087 4 \n",
      "18087 genes with at least 10 reads in sum\n",
      "18087 genes with at least 0 reads per sample\n",
      "18087 genes with at least 0 reads for min 2 sample\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stop1\n",
      "stop2\n",
      "saving all genes to csv...\n",
      "saving to  diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1-all-genes.xls and diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1-all-genes.csv \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in write.table(x = deseqRes, file = resTableCSVFile, quote = F, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqResTPM, file = resTableCSVFileTPM, quote = F, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqResTPMrlog, file = resTableCSVFileRlogTPM, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqRes_signif, file = resTableCSVFile, quote = F, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqResTPM_signif, file = resTableCSVFileTPM, :\n",
      "“appending column names to file”\n",
      "Warning message in write.table(x = deseqResTPMrlog_signif, file = resTableCSVFileRlogTPM, :\n",
      "“appending column names to file”\n"
     ]
    }
   ],
   "source": [
    "head(counts)\n",
    "# n.b. counts from featureCount may not be integers, \n",
    "#  because multi-overlaps are counted as fractions  -> simply round, convert and rebuild matrix\n",
    "\n",
    "allResults <- list() # all results \n",
    "comp_nb <- 1\n",
    "pngNamesVolcano<- NULL\n",
    "comparisonNames <- NULL\n",
    "\n",
    "sumDeseqCalls <- sum(sapply(deseqComparisons, function(x) ncol(x)))\n",
    "\n",
    "sampleAnnot\n",
    "deseqVar <- names(deseqComparisons)[1]\n",
    "for (deseqVar in names(deseqComparisons)) {\n",
    "  head(counts)\n",
    "  if(typeof(counts)!=\"integer\") {\n",
    "    ddsFull = DESeqDataSetFromMatrix( countData = round(counts),colData = sampleAnnot , \n",
    "                                      design = as.formula(paste(\"~\", deseqVar)))\n",
    "  } else {\n",
    "    ddsFull = DESeqDataSetFromMatrix( countData = counts,colData = sampleAnnot , \n",
    "                                      design = as.formula(paste(\"~\", deseqVar)))\n",
    "  }\n",
    "  \n",
    "  # no sample for one factor?\n",
    "  if(length(table(sampleAnnot[,deseqVar]))<2) {\n",
    "    cat(\"skipping\",deseqVar,\", because no sample for one type/factor\\n\")\n",
    "    next\n",
    "  }\n",
    "  \n",
    "  deseqCompCol <- 1\n",
    "  for(deseqCompCol in 1:ncol(deseqComparisons[[deseqVar]])) {\n",
    "    cat(comp_nb, \"of\",sumDeseqCalls,\" - \",paste(deseqVar,\":\",paste(deseqComparisons[deseqVar][[1]][,deseqCompCol],\n",
    "                                                                   collapse=\" vs \"),sep=\"\"),\"\\n\")\n",
    "    comp_nb <- comp_nb+1\n",
    "    compName <- paste(deseqVar,\"_\",paste(deseqComparisons[deseqVar][[1]][,deseqCompCol],collapse=\"_vs_\"),sep=\"\")\n",
    "    # replace / with _:\n",
    "    compName <- gsub(\"\\\\/\",\"_\",compName)\n",
    "\n",
    "    # reduce to requested samples:\n",
    "    colnames(ddsFull)\n",
    "\n",
    "    dds <- ddsFull[,which(sampleAnnot[,deseqVar] %in% deseqComparisons[deseqVar][[1]][,deseqCompCol])]\n",
    "    dim(dds)\n",
    "    head(colData(dds))\n",
    "    colData(dds)[,deseqVar] <- droplevels(colData(dds)[,deseqVar]) # remove non-used factor levels\n",
    "    cat(\"dim of reduced dds\",dim(dds[ rowSums( counts(dds) ) >= minReadCountSum , ]),\"\\n\")\n",
    "    \n",
    "    head(counts(dds))\n",
    "    \n",
    "    # remove genes with 0 or 1 counts for all samples:\n",
    "    dds <- dds[ rowSums( counts(dds) ) >= minReadCountSum , ]\n",
    "    cat(nrow(dds),\"genes with at least\",minReadCountSum,\"reads in sum\\n\")\n",
    "    wh <- which(rowMin(counts(dds)) >= minReadCountPerSample)\n",
    "    if(length(wh)>0) {\n",
    "      dds <- dds[ wh, ]\n",
    "      cat(nrow(dds),\"genes with at least\",minReadCountPerSample,\"reads per sample\\n\")\n",
    "    } else {\n",
    "      cat(\"0 genes left after min read count per sample filter. Won't erase none.\\n\")\n",
    "    }\n",
    "    \n",
    "    # genes with min read counts in two samples\n",
    "    wh <- which(apply(counts(dds), 1,function(r) min(sort(r,decreasing = T)[1:2]))>= minReadCountFor2Samples)\n",
    "    if(length(wh)>0) {\n",
    "      dds <- dds[ wh, ]\n",
    "      cat(nrow(dds),\"genes with at least\",minReadCountFor2Samples,\"reads for min 2 sample\\n\")\n",
    "    } else {\n",
    "      cat(\"0 genes left after min read count per sample filter. Won't erase none.\\n\")\n",
    "    }\n",
    "    \n",
    "    # re-order: normal is the first level -> foldChange right order:\n",
    "    colData(dds)[,deseqVar] <- relevel(colData(dds)[,deseqVar],deseqComparisons[deseqVar][[1]][1,deseqCompCol])\n",
    "    \n",
    "    # make DGE via linear model fitting:\n",
    "    dds <- DESeq(dds)\n",
    "\n",
    "    deseqRes <- as.data.frame(results(dds))\n",
    "    head(deseqRes)\n",
    "    dim(deseqRes)\n",
    "    sampleRowNbs <- ncol(deseqRes)+1\n",
    "    \n",
    "    # add counts:\n",
    "    \n",
    "    # order by group and then alphabetically by id:\n",
    "    id_ordered <- as.character(c( sort(sampleAnnot$ID[which(sampleAnnot[,deseqVar] %in% deseqComparisons[deseqVar][[1]][,deseqCompCol][2])]), \n",
    "                                  sort(sampleAnnot$ID[which(sampleAnnot[,deseqVar] %in% deseqComparisons[deseqVar][[1]][,deseqCompCol][1])])))\n",
    "    head(as.data.frame(assay(dds))[rownames(deseqRes),id_ordered])\n",
    "    deseqRes <- cbind(deseqRes,as.data.frame(assay(dds))[rownames(deseqRes),id_ordered])\n",
    "    sampleRowNbs <- c(sampleRowNbs,ncol(deseqRes))\n",
    "    head(deseqRes)\n",
    "    allResults <- c(allResults,list(deseqRes))\n",
    "    names(allResults)[length(allResults)] <- compName\n",
    "    comparisonNames <- c(comparisonNames,compName)\n",
    "\n",
    "    # add mean, median, sd for all samples and for each group:\n",
    "    head(deseqRes)\n",
    "    s1 <- as.character(sampleAnnot$ID[which(sampleAnnot[,deseqVar] %in% deseqComparisons[deseqVar][[1]][,deseqCompCol][2])])\n",
    "    n1 <- deseqComparisons[deseqVar][[1]][,deseqCompCol][2]\n",
    "\n",
    "    s2 <- as.character(sampleAnnot$ID[which(sampleAnnot[,deseqVar] %in% deseqComparisons[deseqVar][[1]][,deseqCompCol][1])])\n",
    "    n2 <- deseqComparisons[deseqVar][[1]][,deseqCompCol][1]\n",
    "\n",
    "    cat(\"stop1\\n\")\n",
    "\n",
    "    deseqRes[,paste0(n1,\"_mean\")] <- rowMeans(deseqRes[,s1,drop=F])\n",
    "    cat(\"stop2\\n\")\n",
    "    deseqRes[,paste0(n1,\"_median\")] <- apply(deseqRes[,s1,drop=F],1,median)\n",
    "    deseqRes[,paste0(n1,\"_sd\")] <- apply(deseqRes[,s1,drop=F],1,sd)\n",
    "    deseqRes[,paste0(n2,\"_mean\")] <- rowMeans(deseqRes[,s2,drop=F])\n",
    "    deseqRes[,paste0(n2,\"_median\")] <- apply(deseqRes[,s2,drop=F],1,median)\n",
    "    deseqRes[,paste0(n2,\"_sd\")] <- apply(deseqRes[,s2,drop=F],1,sd)\n",
    "    deseqRes[,\"all_mean\"] <- rowMeans(deseqRes[,c(s1,s2),drop=F])\n",
    "    deseqRes[,\"all_median\"] <- apply(deseqRes[,c(s1,s2),drop=F],1,median)\n",
    "    deseqRes[,\"all_sd\"] <- apply(deseqRes[,c(s1,s2),drop=F],1,sd)\n",
    "    \n",
    "    head(deseqRes)\n",
    "    \n",
    "    # same for TMPs:\n",
    "    head(TPMs)\n",
    "    deseqResTPM <- deseqRes\n",
    "    head(rownames(deseqRes))\n",
    "    deseqResTPM[,c(s1,s2)] <- TPMs[rownames(deseqRes),c(s1,s2),drop=F]\n",
    "    deseqResTPM[,paste0(n1,\"_mean\")] <- rowMeans(deseqResTPM[,s1,drop=F])\n",
    "    deseqResTPM[,paste0(n1,\"_median\")] <- apply(deseqResTPM[,s1,drop=F],1,median)\n",
    "    deseqResTPM[,paste0(n1,\"_sd\")] <- apply(deseqResTPM[,s1,drop=F],1,sd)\n",
    "    deseqResTPM[,paste0(n2,\"_mean\")] <- rowMeans(deseqResTPM[,s2,drop=F])\n",
    "    deseqResTPM[,paste0(n2,\"_median\")] <- apply(deseqResTPM[,s2,drop=F],1,median)\n",
    "    deseqResTPM[,paste0(n2,\"_sd\")] <- apply(deseqResTPM[,s2,drop=F],1,sd)\n",
    "    deseqResTPM[,\"all_mean\"] <- rowMeans(deseqResTPM[,c(s1,s2),drop=F])\n",
    "    deseqResTPM[,\"all_median\"] <- apply(deseqResTPM[,c(s1,s2),drop=F],1,median)\n",
    "    deseqResTPM[,\"all_sd\"] <- apply(deseqResTPM[,c(s1,s2),drop=F],1,sd)\n",
    "    tail(deseqResTPM)\n",
    "\n",
    "    # same for rlog TMPs:\n",
    "\t # variance stabilization / rlog transormation: \n",
    "    rld <- varianceStabilizingTransformation(assay(dds))\n",
    "#    rld <- rlog(dds)\n",
    "#    rld <- assay(rld)\n",
    "    head(rld)\n",
    "    deseqResTPMrlog <- deseqRes\n",
    "    deseqResTPMrlog[,c(s1,s2)] <- rld[rownames(deseqRes),c(s1,s2),drop=F]\n",
    "    deseqResTPMrlog[,paste0(n1,\"_mean\")] <- rowMeans(deseqResTPMrlog[,s1,drop=F])\n",
    "    deseqResTPMrlog[,paste0(n1,\"_median\")] <- apply(deseqResTPMrlog[,s1,drop=F],1,median)\n",
    "    deseqResTPMrlog[,paste0(n1,\"_sd\")] <- apply(deseqResTPMrlog[,s1,drop=F],1,sd)\n",
    "    deseqResTPMrlog[,paste0(n2,\"_mean\")] <- rowMeans(deseqResTPMrlog[,s2,drop=F])\n",
    "    deseqResTPMrlog[,paste0(n2,\"_median\")] <- apply(deseqResTPMrlog[,s2,drop=F],1,median)\n",
    "    deseqResTPMrlog[,paste0(n2,\"_sd\")] <- apply(deseqResTPMrlog[,s2,drop=F],1,sd)\n",
    "    deseqResTPMrlog[,\"all_mean\"] <- rowMeans(deseqResTPMrlog[,c(s1,s2),drop=F])\n",
    "    deseqResTPMrlog[,\"all_median\"] <- apply(deseqResTPMrlog[,c(s1,s2),drop=F],1,median)\n",
    "    deseqResTPMrlog[,\"all_sd\"] <- apply(deseqResTPMrlog[,c(s1,s2),drop=F],1,sd)\n",
    "    head(deseqResTPMrlog)\n",
    "    \n",
    "    # save test results of all genes\n",
    "    save(deseqRes,deseqResTPM,deseqResTPMrlog,file = paste(outDir,compName,\"-all-genes-DeSeq2.RData\",sep=\"\"))\n",
    "    # if (saveAllRes) {\n",
    "    if (T) { # always save all results, it's fast enough now, else, change adding gene column below\n",
    "      head(deseqRes)\n",
    "      # save all genes + add annotation:\n",
    "      cat(\"saving all genes to csv...\\n\")\n",
    "      # add annotation:\n",
    "      deseqRes$EnsgID <- rownames(deseqRes)\n",
    "      head(annot)\n",
    "      deseqRes$symbol <- annot[ gsub(\"(\\\\w+).*\",\"\\\\1\",rownames(deseqRes)) ,]$symbol\n",
    "      deseqRes$geneType <- annot[ gsub(\"(\\\\w+).*\",\"\\\\1\",rownames(deseqRes)) ,]$geneType\n",
    "      deseqRes$descr <- annot[ gsub(\"(\\\\w+).*\",\"\\\\1\",rownames(deseqRes)) ,]$description\n",
    "      head(deseqRes)\n",
    "      # add gene symbol as first column again -> fixed first column in Excel\n",
    "      deseqRes$gene <- deseqRes$EnsgID\n",
    "      wh <- which(!is.na(deseqRes$symbol))\n",
    "      if(length(wh)>0){\n",
    "        deseqRes$gene[wh] <- deseqRes$symbol[wh]\n",
    "      }\n",
    "      wh <- which(is.na(deseqRes$symbol))\n",
    "      if(length(wh)>0){\n",
    "        deseqRes$symbol[wh] <- deseqRes$EnsgID[wh]\n",
    "      }\n",
    "      wh <- which(is.na(deseqRes$descr)); if (length(wh)>0) { deseqRes$descr[wh] <- \"\" }\n",
    "      wh <- which(is.na(deseqRes$symbol)); if (length(wh)>0) { deseqRes$symbol[wh] <- \"\" }\n",
    "      wh <- which(is.na(deseqRes$geneType)); if (length(wh)>0) { deseqRes$geneType[wh] <- \"\" }\n",
    "\n",
    "      head(deseqResTPMrlog); #tail(deseqResTPMrlog); tail(deseqResTPMrlog$descr,200)\n",
    "      head(deseqRes); tail(deseqRes); tail(deseqRes$descr,200)\n",
    "      deseqResTPM <- cbind(deseqResTPM,deseqRes[,c(\"gene\",\"EnsgID\",\"symbol\",\"geneType\",\"descr\")])\n",
    "      deseqResTPMrlog <- cbind(deseqResTPMrlog,deseqRes[,c(\"gene\",\"EnsgID\",\"symbol\",\"geneType\",\"descr\")])\n",
    "      \n",
    "      # swap columns to more sensible order:\n",
    "      newColNames <- c(\"gene\",\"EnsgID\",\"geneType\",\"descr\",\"log2FoldChange\",\"padj\",\"pvalue\")\n",
    "      newColNames <- c(newColNames,setdiff(grep(\"mean\",colnames(deseqRes),value = T),\"all_mean\"),\"baseMean\", \"lfcSE\",\"stat\")\n",
    "      newColNames <- c(newColNames,setdiff(setdiff(colnames(deseqRes),newColNames),id_ordered),id_ordered)\n",
    "      deseqRes <- deseqRes[,newColNames]\n",
    "      deseqResTPM <- deseqResTPM[,newColNames]\n",
    "      deseqResTPMrlog <- deseqResTPMrlog[,newColNames]\n",
    "      head(deseqResTPM); tail(deseqResTPM)\n",
    "      head(unique(deseqResTPMrlog$descr))\n",
    "      # save as xls\n",
    "    \n",
    "      resTableCSVFile <- paste(outDir,compName,\"-all-genes.csv\",sep=\"\")\n",
    "      resTableXLSFile <- paste(outDir,compName,\"-all-genes.xls\",sep=\"\")\n",
    "      resTableCSVFileTPM <- paste(outDir,compName,\"-all-genes-tpm.csv\",sep=\"\")\n",
    "      resTableCSVFileRlogTPM <- paste(outDir,compName,\"-all-genes-rlogTpm.csv\",sep=\"\")\n",
    "\n",
    "      cat(\"saving to \",resTableXLSFile, \"and\",resTableCSVFile,\"\\n\")\n",
    "      header <- rep(\"\",ncol(deseqRes))\n",
    "      wh <- which(colnames(deseqRes) %in% sampleAnnot$ID)\n",
    "      head(sampleAnnot)\n",
    "      header[wh] <-  as.character(sampleAnnot[colnames(deseqRes)[wh],deseqVar])\n",
    "      write.table(x = t(header),file = resTableCSVFile,quote = F,sep = \"\\t\",row.names = F, col.names = F)\n",
    "      write.table(x = deseqRes,file = resTableCSVFile,quote = F,sep = \"\\t\",row.names = F, append=T)\n",
    "      write.table(x = t(header),file = resTableCSVFileTPM,quote = F,sep = \"\\t\",row.names = F, col.names = F)\n",
    "      write.table(x = deseqResTPM,file = resTableCSVFileTPM,quote = F,sep = \"\\t\",row.names = F, append=T)\n",
    "      write.table(x = t(header),file = resTableCSVFileRlogTPM,quote = F,sep = \"\\t\",row.names = F, col.names = F)\n",
    "      write.table(x = deseqResTPMrlog,file = resTableCSVFileRlogTPM,quote = F,sep = \"\\t\",row.names = F, append=T)\n",
    "\n",
    "    # saving xls doesn't currently work\n",
    "    # have to install perl with corresp modules\n",
    "    # better install R library xlsx ... didn't work for me here\n",
    "        \n",
    "        # convert to xls with external script:\n",
    "#      cmd <- paste0(\"perl \",RNAScriptPath,\"/dge-results-csvs2xls-batch.pl \\\"\\\\t\\\" 2 10 \",\n",
    "#                    resTableXLSFile,\" \", resTableCSVFile,\" \", resTableCSVFileTPM,\" \", resTableCSVFileRlogTPM, \" results_counts results_TPMs results_rlogTPMs\")\n",
    "#      cat(cmd,\"\\n\")\n",
    "#      system(command = cmd)\n",
    "#      file.remove(resTableCSVFile,resTableCSVFileTPM,resTableCSVFileRlogTPM)\n",
    "    }\n",
    "\n",
    "    # remove results where padj = NA\n",
    "    wh <- which(is.na(deseqRes$padj))\n",
    "    if (length(wh)>0) { \n",
    "      deseqRes <- deseqRes[ -wh,]\n",
    "      deseqResTPM <- deseqResTPM[-wh,]\n",
    "      deseqResTPMrlog <- deseqResTPMrlog[-wh,]\n",
    "    }\n",
    "    # keep only signif. results (after multi-testing):\n",
    "    if (ncol(dds)>2) {\n",
    "      deseqRes_signif <- deseqRes[ deseqRes$padj<padjCutoff,]\n",
    "      deseqResTPM_signif <- deseqResTPM[ deseqResTPM$padj<padjCutoff,]\n",
    "      deseqResTPMrlog_signif <- deseqResTPMrlog[ deseqResTPMrlog$padj<padjCutoff,]\n",
    "    } else {\n",
    "      # in case there are only 2 samples (no replicates), there keep all results to keep FoldChanges of all\n",
    "      deseqRes_signif <- deseqRes\n",
    "      deseqResTPM_signif <- deseqResTPM\n",
    "      deseqResTPMrlog_signif <- deseqResTPMrlog\n",
    "    }\n",
    "    \n",
    "    deseqRes_signif <- deseqRes_signif[ order(deseqRes_signif$padj),]\n",
    "    deseqResTPM_signif <- deseqResTPM_signif[ order(deseqResTPM_signif$padj),]\n",
    "    deseqResTPMrlog_signif <- deseqResTPMrlog_signif[ order(deseqResTPMrlog_signif$padj),]\n",
    "    \n",
    "    head(deseqRes_signif)\n",
    "    tail(deseqRes_signif)\n",
    "    dim(deseqRes_signif)\n",
    "\n",
    "    # save results:\n",
    "  \n",
    "    #write.table(deseqRes_signif,paste(outDir,compName,\"-signif-DGE-genes.csv\",sep=\"\"),quote = F,row.names = F,sep=\"\\t\", dec=\".\")\n",
    "    save(deseqRes_signif,file = paste(outDir,compName,\"-signif-DGE-genes.RData\",sep=\"\"))\n",
    "    \n",
    "    # save as csv:\n",
    "    resTableCSVFile <- paste(outDir,compName,\"-signif-DEGs.csv\",sep=\"\")\n",
    "    resTableXLSFile <- paste(outDir,compName,\"-signif-DEGs.xls\",sep=\"\")\n",
    "    resTableCSVFileTPM <- paste(outDir,compName,\"-signif-DEGs-tpm.csv\",sep=\"\")\n",
    "    resTableCSVFileRlogTPM <- paste(outDir,compName,\"-signif-DEGs-rlogTpm.csv\",sep=\"\")\n",
    "    write.table(x = t(header),file = resTableCSVFile,quote = F,sep = \"\\t\",row.names = F, col.names = F)\n",
    "    write.table(x = deseqRes_signif,file = resTableCSVFile,quote = F,sep = \"\\t\",row.names = F, append=T)\n",
    "    write.table(x = t(header),file = resTableCSVFileTPM,quote = F,sep = \"\\t\",row.names = F, col.names = F)\n",
    "    write.table(x = deseqResTPM_signif,file = resTableCSVFileTPM,quote = F,sep = \"\\t\",row.names = F, append=T)\n",
    "    write.table(x = t(header),file = resTableCSVFileRlogTPM,quote = F,sep = \"\\t\",row.names = F, col.names = F)\n",
    "    write.table(x = deseqResTPMrlog_signif,file = resTableCSVFileRlogTPM,quote = F,sep = \"\\t\",row.names = F, append=T)\n",
    "    # convert to xls with external script:\n",
    "#    cmd <- paste0(\"perl \",RNAScriptPath,\"/dge-results-csvs2xls-batch.pl \\\"\\\\t\\\" 2 10 \",\n",
    "#                  resTableXLSFile,\" \", resTableCSVFile,\" \", resTableCSVFileTPM,\" \", resTableCSVFileRlogTPM, \" results_counts results_TPMs results_rlogTPMs\")\n",
    "#    cat(cmd,\"\\n\")\n",
    "#    system(command = cmd)\n",
    "#    file.remove(resTableCSVFile,resTableCSVFileTPM,resTableCSVFileRlogTPM)\n",
    "\n",
    "    # save single genes of interest:\n",
    "    if (extraEnsgIDFile!=\"n\") {\n",
    "          \n",
    "      deseqRes_singleGenes <- deseqRes\n",
    "      deseqRes_singleGenes <- deseqRes_singleGenes[extraEnsgID,]\n",
    "      # add annotation:\n",
    "      deseqRes_singleGenes$EnsgID <- rownames(deseqRes_singleGenes)\n",
    "      head(annot)\n",
    "      deseqRes_singleGenes$symbol <- annot[ gsub(\"(\\\\w+).*\",\"\\\\1\",rownames(deseqRes_singleGenes)) ,]$symbol\n",
    "      deseqRes_singleGenes$geneType <- annot[ gsub(\"(\\\\w+).*\",\"\\\\1\",rownames(deseqRes_singleGenes)) ,]$geneType\n",
    "      deseqRes_singleGenes$descr <- annot[ gsub(\"(\\\\w+).*\",\"\\\\1\",rownames(deseqRes_singleGenes)) ,]$description\n",
    "      write.table(deseqRes_singleGenes,paste(outDir,compName,\"-single-genes.csv\",sep=\"\"),quote = F,row.names = F,sep=\"\\t\", dec=\".\")\n",
    "    }  \n",
    "\n",
    "    # plot     \n",
    "    exponents <- -log10(deseqRes$pvalue)\n",
    "    maxEpo <- max(exponents[exponents!=Inf])\n",
    "    deseqRes$threshold <- ifelse( deseqRes$padj < pCutoff,\"Benjamini-Hochberg\",\"unsignificant\")\n",
    "    \n",
    "    # volcano plot\n",
    "    g <- ggplot(data=deseqRes, aes(x=log2FoldChange, y=-log10(pvalue), colour=threshold)) +\n",
    "      geom_point(alpha=0.4, size=1.25) +\n",
    "      #  opts(legend.position = \"none\") +\n",
    "      xlim(c(-10, 10)) + ylim(c(0, maxEpo)) +\n",
    "      xlab(\"log2 fold change\") + ylab(\"-log10 p-value\") +\n",
    "      ggtitle(paste(compName,\"\\nvolcano plot\"))\n",
    "    ggsave(filename = paste(outDir,compName,\"-volcano.png\",sep=\"\"),width = 7, height=7,dpi = 150)\n",
    "    pngNamesVolcano <- c(pngNamesVolcano,paste(outDir,compName,\"-volcano.png\",sep=\"\"))\n",
    "    \n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a6c3ec35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/gruetzmko/miniconda3/envs/imagemagick/bin/convert diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT913-volcano.png diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1-volcano.png diff-gene-exp/volcano-plots-all.pdf\n",
      " \n"
     ]
    }
   ],
   "source": [
    "# merge Volcano plots:\n",
    "cmd = paste(imageMagickConvert,\" \",paste(pngNamesVolcano,collapse=\" \"),\" \",\n",
    "          outDir,\"volcano-plots-all.pdf\\n\",sep=\"\")\n",
    "cat(cmd,\"\\n\")\n",
    "system(cmd)\n",
    "system(paste(\"rm \",paste(pngNamesVolcano,collapse=\" \"),sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "98a5f0cd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GeneID</th><th scope=col>Chr</th><th scope=col>Start</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000223972</th><td>ENSG00000223972</td><td>chr1</td><td>11869</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000227232</th><td>ENSG00000227232</td><td>chr1</td><td>14363</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000243485</th><td>ENSG00000243485</td><td>chr1</td><td>29554</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237613</th><td>ENSG00000237613</td><td>chr1</td><td>34554</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000268020</th><td>ENSG00000268020</td><td>chr1</td><td>52473</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000240361</th><td>ENSG00000240361</td><td>chr1</td><td>62948</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & GeneID & Chr & Start\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tENSG00000223972 & ENSG00000223972 & chr1 & 11869\\\\\n",
       "\tENSG00000227232 & ENSG00000227232 & chr1 & 14363\\\\\n",
       "\tENSG00000243485 & ENSG00000243485 & chr1 & 29554\\\\\n",
       "\tENSG00000237613 & ENSG00000237613 & chr1 & 34554\\\\\n",
       "\tENSG00000268020 & ENSG00000268020 & chr1 & 52473\\\\\n",
       "\tENSG00000240361 & ENSG00000240361 & chr1 & 62948\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | GeneID &lt;chr&gt; | Chr &lt;chr&gt; | Start &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| ENSG00000223972 | ENSG00000223972 | chr1 | 11869 |\n",
       "| ENSG00000227232 | ENSG00000227232 | chr1 | 14363 |\n",
       "| ENSG00000243485 | ENSG00000243485 | chr1 | 29554 |\n",
       "| ENSG00000237613 | ENSG00000237613 | chr1 | 34554 |\n",
       "| ENSG00000268020 | ENSG00000268020 | chr1 | 52473 |\n",
       "| ENSG00000240361 | ENSG00000240361 | chr1 | 62948 |\n",
       "\n"
      ],
      "text/plain": [
       "                GeneID          Chr  Start\n",
       "ENSG00000223972 ENSG00000223972 chr1 11869\n",
       "ENSG00000227232 ENSG00000227232 chr1 14363\n",
       "ENSG00000243485 ENSG00000243485 chr1 29554\n",
       "ENSG00000237613 ENSG00000237613 chr1 34554\n",
       "ENSG00000268020 ENSG00000268020 chr1 52473\n",
       "ENSG00000240361 ENSG00000240361 chr1 62948"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"group_full_UWB1.289_ctr_vs_UWB1.289_CT913\"\n",
      "[1] \"group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1\"\n",
      "[1] \"group_full_UWB1.289_ctr_vs_UWB1.289_CT913\"\n",
      "[1] \"group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1\"\n"
     ]
    }
   ],
   "source": [
    "# # manhattan plots    png\n",
    "head(genePos)\n",
    "pngNames <- NULL\n",
    "comp_nb <- 1\n",
    "plotValue <- \"pvalue\"\n",
    "for (plotValue in c(\"pvalue\",\"log2FoldChange\")) {\n",
    "  for (comp_nb in 1:length(allResults)) {\n",
    "    #for (comp_nb in 1:2) {\n",
    "    pngNames <- c(pngNames,paste(outDir,names(allResults)[comp_nb],\"-manhattan-plots-all-\",plotValue,\".png\",sep=\"\"))\n",
    "    png(width = 1200, height = 600, file = paste(outDir,names(allResults)[comp_nb],\"-manhattan-plots-all-\",plotValue,\".png\",sep=\"\"), res=100)\n",
    "    print(names(allResults)[comp_nb])\n",
    "    res <- allResults[comp_nb][[1]]\n",
    "    res$EnsgID <- rownames(res)\n",
    "    # head(genePos)\n",
    "    # head(rownames(res))\n",
    "    # table(nchar(rownames(genePos)))\n",
    "    # table(nchar(rownames(res)))\n",
    "    res$Chr <- genePos[rownames(res),\"Chr\"]\n",
    "    if(length(grep(\"^chr\",res$Chr))==0) {\n",
    "      res$Chr <- paste0(\"chr\",res$Chr)\n",
    "    }\n",
    "    res$Start <- genePos[rownames(res),\"Start\"]\n",
    "    res$Start <- as.numeric(res$Start)\n",
    "    res$ChrNum <- gsub(\"^chr\",\"\",res$Chr)\n",
    "    # add numbers for the unnumerated chr:\n",
    "    head(res$ChrNum)\n",
    "    unique(res$ChrNum)\n",
    "    wh <- grep(\"^\\\\d\",x = res$ChrNum)\n",
    "    if(length(res$ChrNum)-length(wh) > 0) {\n",
    "      maxNb <- max(as.numeric(res$ChrNum[wh]))\n",
    "      repl <- 1:length(unique(res$ChrNum[-wh]))+maxNb\n",
    "      names(repl) <- unique(res$ChrNum[-wh])\n",
    "      res$ChrNum[-wh] <- repl[res$ChrNum[-wh]]\n",
    "    }\n",
    "    unique(res$ChrNum)\n",
    "    unique(res$Chr)\n",
    "    res$ChrNum <- as.numeric(res$ChrNum)\n",
    "\n",
    "    # omit chr that start without number\n",
    "    wh <- grep(\"^chr\",res$Chr, invert=T)\n",
    "    unique(res$Chr[wh])\n",
    "    if (length(wh)>0) { res <- res[ -wh,]}\n",
    "    head(res)\n",
    "    table(res$Chr)\n",
    "    table(res$ChrNum)\n",
    "\n",
    "    # sort results by chrNum:\n",
    "    res <- res[order(res$ChrNum),]\n",
    "    \n",
    "    genomeWideLineVal <- -log10(pCutoff/nrow(res))\n",
    "    res <- res[!is.na(res$pvalue),]\n",
    "    whHighlight <- res$EnsgID[which(res$padj<padjCutoff & abs(res$log2FoldChange)>2)]\n",
    "    nbSignif <- length(which(res$padj<0.05))\n",
    "    \n",
    "    y2lim <- -log10(min(res$pvalue[res$pvalue!=0]))\n",
    "    if (plotValue==\"pvalue\") {\n",
    "      manhattan(res,chr = \"ChrNum\",bp = \"Start\",p = \"pvalue\",snp = \"EnsgID\",genomewideline = genomeWideLineVal, suggestiveline=F,\n",
    "                highlight = whHighlight, col = c(\"blue\",\"gold\"),main=paste0(plotValue,\" - \",names(allResults)[comp_nb]),\n",
    "                ylim=c(0,y2lim), chrlabs=unique(res$Chr), cex = 0.6, cex.lab=0.6, cex.axis=0.6)\n",
    "      text(x = par(\"usr\")[2]*1.03,y = genomeWideLineVal*0.97,labels = \"Bonferroni\\ncutoff\",srt=0,xpd=NA,cex=0.8, pos = 2)\n",
    "    } else {\n",
    "      if (plotValue==\"log2FoldChange\") {\n",
    "        genomeWideLineVal<-0\n",
    "        manhattan(res,chr = \"ChrNum\",bp = \"Start\",p = \"log2FoldChange\",snp = \"EnsgID\",genomewideline = genomeWideLineVal, suggestiveline=F,\n",
    "                  highlight = whHighlight, col = c(\"#0e6655\",\"#f5b041\"),main=paste0(plotValue,\" - \",names(allResults)[comp_nb]),\n",
    "                  ylim=c(min(res$log2FoldChange),max(res$log2FoldChange)),\n",
    "                  chrlabs=unique(res$Chr), cex = 0.6, cex.lab=0.6, cex.axis=0.6,  logp=F, ylab=expression(\"log\"[2]*\"(FC)\"))\n",
    "        abline(h = 0,col=\"red\")\n",
    "      }\n",
    "    }\n",
    "\n",
    "    mtext(text = paste(nbSignif,\"genes with p-adjusted < 0.05\\ngreen: padj<0.05 & abs(log2FC) >2  -> \",\n",
    "                       length(whHighlight),\"genes\"),side = 3,cex = 0.8, padj=0.1)\n",
    "    dev.off()\n",
    "  }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "729cb4a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/gruetzmko/miniconda3/envs/imagemagick/bin/convert diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT913-manhattan-plots-all-pvalue.png diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1-manhattan-plots-all-pvalue.png diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT913-manhattan-plots-all-log2FoldChange.png diff-gene-exp/group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1-manhattan-plots-all-log2FoldChange.png diff-gene-exp/manhattan-plots-all.pdf\n",
      " \n"
     ]
    }
   ],
   "source": [
    "# merge manhattan png's:\n",
    "\n",
    "cmd = paste(imageMagickConvert,\" \",paste(pngNames,collapse=\" \"),\" \",\n",
    "            outDir,\"manhattan-plots-all.pdf\\n\",sep=\"\")\n",
    "cat(cmd,\"\\n\")\n",
    "system(cmd)\n",
    "#system(paste(\"rm \",paste(pngNames,collapse=\" \"),sep=\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f203dfb",
   "metadata": {},
   "source": [
    "### overview statistics:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "377a6472",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fold change cutoffs:  0 1 1.5 2 3 5"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 19</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>comparison</th><th scope=col>all</th><th scope=col>up</th><th scope=col>down</th><th scope=col>all</th><th scope=col>up</th><th scope=col>down</th><th scope=col>all</th><th scope=col>up</th><th scope=col>down</th><th scope=col>all</th><th scope=col>up</th><th scope=col>down</th><th scope=col>all</th><th scope=col>up</th><th scope=col>down</th><th scope=col>all</th><th scope=col>up</th><th scope=col>down</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>group_full_UWB1.289_ctr_vs_UWB1.289_CT913</th><td>group_full_UWB1.289_ctr_vs_UWB1.289_CT913</td><td>6574</td><td>3098</td><td>3476</td><td>1606</td><td>864</td><td>742</td><td>668</td><td>351</td><td>317</td><td>331</td><td>174</td><td>157</td><td>92</td><td>56</td><td>36</td><td>8</td><td>4</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1</th><td>group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1</td><td>3441</td><td>1499</td><td>1942</td><td> 411</td><td>212</td><td>199</td><td>120</td><td> 62</td><td> 58</td><td> 34</td><td> 16</td><td> 18</td><td> 4</td><td> 4</td><td> 0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 19\n",
       "\\begin{tabular}{r|lllllllllllllllllll}\n",
       "  & comparison & all & up & down & all & up & down & all & up & down & all & up & down & all & up & down & all & up & down\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tgroup\\_full\\_UWB1.289\\_ctr\\_vs\\_UWB1.289\\_CT913 & group\\_full\\_UWB1.289\\_ctr\\_vs\\_UWB1.289\\_CT913 & 6574 & 3098 & 3476 & 1606 & 864 & 742 & 668 & 351 & 317 & 331 & 174 & 157 & 92 & 56 & 36 & 8 & 4 & 4\\\\\n",
       "\tgroup\\_full\\_UWB1.289\\_ctr\\_vs\\_UWB1.289\\_CT\\_M1 & group\\_full\\_UWB1.289\\_ctr\\_vs\\_UWB1.289\\_CT\\_M1 & 3441 & 1499 & 1942 &  411 & 212 & 199 & 120 &  62 &  58 &  34 &  16 &  18 &  4 &  4 &  0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 19\n",
       "\n",
       "| <!--/--> | comparison &lt;chr&gt; | all &lt;dbl&gt; | up &lt;dbl&gt; | down &lt;dbl&gt; | all &lt;dbl&gt; | up &lt;dbl&gt; | down &lt;dbl&gt; | all &lt;dbl&gt; | up &lt;dbl&gt; | down &lt;dbl&gt; | all &lt;dbl&gt; | up &lt;dbl&gt; | down &lt;dbl&gt; | all &lt;dbl&gt; | up &lt;dbl&gt; | down &lt;dbl&gt; | all &lt;dbl&gt; | up &lt;dbl&gt; | down &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| group_full_UWB1.289_ctr_vs_UWB1.289_CT913 | group_full_UWB1.289_ctr_vs_UWB1.289_CT913 | 6574 | 3098 | 3476 | 1606 | 864 | 742 | 668 | 351 | 317 | 331 | 174 | 157 | 92 | 56 | 36 | 8 | 4 | 4 |\n",
       "| group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1 | group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1 | 3441 | 1499 | 1942 |  411 | 212 | 199 | 120 |  62 |  58 |  34 |  16 |  18 |  4 |  4 |  0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "                                          comparison                               \n",
       "group_full_UWB1.289_ctr_vs_UWB1.289_CT913 group_full_UWB1.289_ctr_vs_UWB1.289_CT913\n",
       "group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1 group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1\n",
       "                                          all  up   down all  up  down all up \n",
       "group_full_UWB1.289_ctr_vs_UWB1.289_CT913 6574 3098 3476 1606 864 742  668 351\n",
       "group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1 3441 1499 1942  411 212 199  120  62\n",
       "                                          down all up  down all up down all up\n",
       "group_full_UWB1.289_ctr_vs_UWB1.289_CT913 317  331 174 157  92  56 36   8   4 \n",
       "group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1  58   34  16  18   4   4  0   0   0 \n",
       "                                          down\n",
       "group_full_UWB1.289_ctr_vs_UWB1.289_CT913 4   \n",
       "group_full_UWB1.289_ctr_vs_UWB1.289_CT_M1 0   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "comp_nb<-1\n",
    "l2fcVas <- c(0,1,1.5,2,3,5)\n",
    "metaColNames <-as.character(sapply(l2fcVas,function(x) c(paste(\"abs(log2FC) >\",x),\"\",\"\")))\n",
    "colNames <- as.character(sapply(1:length(l2fcVas), function(x) c(\"all\",\"up\",\"down\")))\n",
    "ovStat <- matrix(rep(0,length(allResults)*length(l2fcVas)*3),nrow=length(allResults),\n",
    "                 dimnames = list(names(allResults),colNames))\n",
    "for (comp_nb in 1:length(allResults)) {\n",
    "  name <- names(allResults[1])\n",
    "  j<-2\n",
    "  for (j in 1:length(l2fcVas)) {\n",
    "    whL2Fc <- which(abs(allResults[[comp_nb]]$log2FoldChange)>l2fcVas[j] & allResults[[comp_nb]]$padj<padjCutoff)\n",
    "    whL2FcUp <- which(allResults[[comp_nb]]$log2FoldChange>l2fcVas[j] & allResults[[comp_nb]]$padj<padjCutoff) \n",
    "    whL2FcDn <- which(allResults[[comp_nb]]$log2FoldChange < -l2fcVas[j] & allResults[[comp_nb]]$padj<padjCutoff)\n",
    "    ovStat[comp_nb,(j-1)*3+1] <- length(whL2Fc)\n",
    "    ovStat[comp_nb,(j-1)*3+2] <- length(whL2FcUp)\n",
    "    ovStat[comp_nb,(j-1)*3+3] <- length(whL2FcDn)\n",
    "  }\n",
    "}\n",
    "\n",
    "ovStat <- as.data.frame(ovStat)\n",
    "ovStat <- cbind(\"comparison\"=rownames(ovStat),ovStat)\n",
    "ovStat$comparison <- as.character(ovStat$comparison)\n",
    "cat(\"fold change cutoffs: \",l2fcVas)\n",
    "ovStat\n",
    "i <- 1\n",
    "for (deseqVar in names(deseqComparisons)) {\n",
    "  \n",
    "  for(deseqCompCol in 1:ncol(deseqComparisons[[deseqVar]])) {\n",
    "    # requested samples:\n",
    "    l1 <- length(which(sampleAnnot[,deseqVar] %in% deseqComparisons[deseqVar][[1]][,deseqCompCol][1]))\n",
    "    l2 <- length(which(sampleAnnot[,deseqVar] %in% deseqComparisons[deseqVar][[1]][,deseqCompCol][2]))\n",
    "    ovStat$comparison[i] <- paste(ovStat$comparison[i],\" (\",l1,\" vs \",l2,\")\",sep=\"\")\n",
    "    i <-i+1\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "776b68d0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "write.table(x = ovStat, file = overviewStatFile, quote = F,sep = \"\\t\",row.names = F)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
